{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d90955b5-95f2-41f5-bb67-79d913e3569d",
   "metadata": {},
   "source": [
    "## Preliminary Exploration\n",
    "### P1: Using `sentence_transformers` \n",
    "There are many pre-trained models available in the `sentence_transformers` (a.k.a SBERT) library in Python. [ *Reference: https://sbert.net/* ]\n",
    "\n",
    "To make sure the library is installed, run following:\n",
    "<pre>\n",
    "pip install sentence_transformers    \n",
    "</pre>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e720d619-978b-46cc-b327-8bc93454734a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: sentence_transformers in c:\\users\\benlc\\miniforge3\\lib\\site-packages (4.1.0)\n",
      "Requirement already satisfied: transformers<5.0.0,>=4.41.0 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from sentence_transformers) (4.52.4)\n",
      "Requirement already satisfied: tqdm in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from sentence_transformers) (4.66.5)\n",
      "Requirement already satisfied: torch>=1.11.0 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from sentence_transformers) (2.7.1)\n",
      "Requirement already satisfied: scikit-learn in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from sentence_transformers) (1.6.0)\n",
      "Requirement already satisfied: scipy in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from sentence_transformers) (1.14.1)\n",
      "Requirement already satisfied: huggingface-hub>=0.20.0 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from sentence_transformers) (0.33.0)\n",
      "Requirement already satisfied: Pillow in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from sentence_transformers) (11.0.0)\n",
      "Requirement already satisfied: typing_extensions>=4.5.0 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from sentence_transformers) (4.12.2)\n",
      "Requirement already satisfied: filelock in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from huggingface-hub>=0.20.0->sentence_transformers) (3.18.0)\n",
      "Requirement already satisfied: fsspec>=2023.5.0 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from huggingface-hub>=0.20.0->sentence_transformers) (2025.5.1)\n",
      "Requirement already satisfied: packaging>=20.9 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from huggingface-hub>=0.20.0->sentence_transformers) (24.1)\n",
      "Requirement already satisfied: pyyaml>=5.1 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from huggingface-hub>=0.20.0->sentence_transformers) (6.0.2)\n",
      "Requirement already satisfied: requests in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from huggingface-hub>=0.20.0->sentence_transformers) (2.32.3)\n",
      "Requirement already satisfied: sympy>=1.13.3 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from torch>=1.11.0->sentence_transformers) (1.14.0)\n",
      "Requirement already satisfied: networkx in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from torch>=1.11.0->sentence_transformers) (3.4.2)\n",
      "Requirement already satisfied: jinja2 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from torch>=1.11.0->sentence_transformers) (3.1.6)\n",
      "Requirement already satisfied: setuptools in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from torch>=1.11.0->sentence_transformers) (75.1.0)\n",
      "Requirement already satisfied: colorama in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from tqdm->sentence_transformers) (0.4.6)\n",
      "Requirement already satisfied: numpy>=1.17 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from transformers<5.0.0,>=4.41.0->sentence_transformers) (2.0.2)\n",
      "Requirement already satisfied: regex!=2019.12.17 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from transformers<5.0.0,>=4.41.0->sentence_transformers) (2024.11.6)\n",
      "Requirement already satisfied: tokenizers<0.22,>=0.21 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from transformers<5.0.0,>=4.41.0->sentence_transformers) (0.21.1)\n",
      "Requirement already satisfied: safetensors>=0.4.3 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from transformers<5.0.0,>=4.41.0->sentence_transformers) (0.5.3)\n",
      "Requirement already satisfied: joblib>=1.2.0 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from scikit-learn->sentence_transformers) (1.4.2)\n",
      "Requirement already satisfied: threadpoolctl>=3.1.0 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from scikit-learn->sentence_transformers) (3.5.0)\n",
      "Requirement already satisfied: mpmath<1.4,>=1.1.0 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from sympy>=1.13.3->torch>=1.11.0->sentence_transformers) (1.3.0)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from jinja2->torch>=1.11.0->sentence_transformers) (3.0.2)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from requests->huggingface-hub>=0.20.0->sentence_transformers) (3.3.2)\n",
      "Requirement already satisfied: idna<4,>=2.5 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from requests->huggingface-hub>=0.20.0->sentence_transformers) (3.10)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from requests->huggingface-hub>=0.20.0->sentence_transformers) (2.2.3)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from requests->huggingface-hub>=0.20.0->sentence_transformers) (2024.8.30)\n"
     ]
    }
   ],
   "source": [
    "!pip install sentence_transformers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7636650e-3015-427f-ac2f-59e48b460794",
   "metadata": {},
   "source": [
    "Now, let's use the `sentence_transformers` library and import a pre-trained model named `all-MiniLM-L6-v2`\n",
    "\n",
    "1. First we import the library using:\n",
    "\n",
    "   `from sentence_transformers import SentenceTransformer`\n",
    "\n",
    "  \n",
    "2. Then we will need to define a variable to hold the pre-trained model:\n",
    "\n",
    "   `embedding_model = SentenceTransformer('all-MiniLM-L6-v2')`\n",
    "\n",
    "   ***Note:** There could be many options for the pre-trained model, such as:*\n",
    "   - `all-MiniLM-L6-v2` which is overall fast and good for general text (Dim: 384)\n",
    "   - `all-mpnet-base-v2` which is slower but better quality (Dim: 768)\n",
    "    \n",
    "   For more details: https://huggingface.co/sentence-transformers\n",
    "\n",
    "3. Once the model is loaded, we can use the `encode()` function within the embedded model to transform the given input text into corresponding vector.\n",
    "\n",
    "    `embedding_model.encode(input_text_here)`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8362c925-8c91-4628-9a7b-8b68e5f0048c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\benlc\\miniforge3\\Lib\\site-packages\\tqdm\\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original text: Today is a sunny day.\n",
      "Vector length: 384\n",
      "Embedding vector: [ 2.36809417e-03  9.73476768e-02  9.59824622e-02  9.13097858e-02\n",
      "  3.10873557e-02 -5.90473823e-02  1.50612637e-01 -9.29311365e-02\n",
      " -3.90396342e-02 -1.75307281e-02  2.53553204e-02  1.69024300e-02\n",
      "  2.84761633e-03  7.10389912e-02  8.43285397e-02  6.34109676e-02\n",
      " -2.40979977e-02 -9.74420982e-04 -6.98793307e-02  2.91991327e-02\n",
      " -1.09861061e-01  9.27486364e-03 -3.49005498e-02  2.89231036e-02\n",
      "  9.71570984e-03  7.08094239e-02  4.00788449e-02 -1.34750886e-03\n",
      " -3.73694114e-02 -2.54107974e-02 -2.10091248e-02  5.22645079e-02\n",
      " -3.07825413e-02 -4.27118735e-03 -1.77656952e-02 -3.46577615e-02\n",
      " -6.61386782e-03 -1.32381096e-01 -4.01914157e-02  1.00402310e-01\n",
      " -5.23531204e-03 -1.35367617e-01  6.83695031e-03  1.96054019e-02\n",
      " -6.42637396e-03 -2.14972515e-02 -3.69780436e-02  6.56473190e-02\n",
      "  1.03695050e-01 -4.45263945e-02 -5.79723977e-02 -4.58558882e-03\n",
      " -2.62573659e-02 -4.00760807e-02 -3.66745777e-02  9.90855992e-02\n",
      "  4.08710763e-02 -4.31849211e-02  9.24060270e-02  3.60976346e-03\n",
      "  1.70968119e-02  2.52813771e-02 -6.04157597e-02  4.24370877e-02\n",
      "  9.51794013e-02 -4.43955436e-02 -8.73884782e-02 -4.15904522e-02\n",
      " -3.59408483e-02 -3.67453061e-02 -7.55787790e-02  2.72148848e-02\n",
      "  3.72199900e-02  1.43180080e-02 -2.28820816e-02 -3.60709578e-02\n",
      "  2.15278734e-02 -6.55717310e-03  3.87540609e-02 -6.92155287e-02\n",
      "  2.49279551e-02 -5.58850616e-02 -4.60230093e-03  3.86204422e-02\n",
      " -6.30975468e-04  2.16699578e-02 -2.75513008e-02  9.34047773e-02\n",
      "  1.85430124e-02 -3.07690576e-02 -1.55948192e-01 -5.18612452e-02\n",
      " -5.37523925e-02  2.15563811e-02 -2.59783957e-02  8.00930057e-03\n",
      " -1.44009516e-02  1.49148637e-02 -4.88745309e-02  8.33512843e-02\n",
      "  4.52757366e-02  2.75515039e-02 -6.59208149e-02  2.53418013e-02\n",
      "  7.71512240e-02  6.06045127e-03 -5.10935523e-02  4.83395457e-02\n",
      " -3.22792977e-02 -6.26896173e-02 -2.45411601e-02  4.52717692e-02\n",
      "  4.10760083e-04  5.57307713e-03  4.07143869e-02 -5.79886846e-02\n",
      "  2.21302267e-02  3.15777846e-02 -9.88128111e-02 -9.52431932e-03\n",
      "  2.85894908e-02  1.31186144e-02  3.51919420e-02  9.84260440e-02\n",
      "  3.67152989e-02 -1.74260903e-02  8.21214914e-02 -7.88951033e-33\n",
      "  8.38996470e-02  5.29275201e-02  2.23860312e-02  3.98712978e-02\n",
      "  4.98515256e-02 -1.54442098e-02 -3.41371186e-02 -2.09566168e-02\n",
      " -1.36216935e-02 -1.08017817e-01 -2.48924140e-02 -8.36118590e-03\n",
      " -4.40845191e-02  1.48956124e-02  2.61315182e-02 -6.83197752e-02\n",
      "  4.04800251e-02  8.46970379e-02 -3.01017333e-02  9.66638327e-02\n",
      " -1.09072797e-01  4.12751622e-02  3.69675048e-02 -5.09032942e-02\n",
      " -6.70546815e-02 -1.99374128e-02  4.81363758e-02 -1.73276849e-02\n",
      " -8.37802654e-05  4.14708164e-03  8.75265524e-02 -5.49884476e-02\n",
      "  9.24860016e-02  6.04895018e-02  3.13180611e-02 -4.86725941e-02\n",
      " -1.28271924e-02 -3.95995006e-02 -1.13016944e-02 -4.83527547e-03\n",
      " -4.97682802e-02  1.17009506e-03  2.40090839e-03 -1.50692808e-02\n",
      " -1.03200125e-02  8.02762061e-02  3.83658744e-02  4.78069410e-02\n",
      " -6.86059743e-02  3.00509520e-02 -5.78049496e-02  2.54423730e-02\n",
      "  1.13577163e-02 -2.21754741e-02 -3.80734317e-02 -4.82005067e-02\n",
      " -1.10568712e-02 -5.95117034e-03 -2.17876825e-02  4.50677760e-02\n",
      " -2.70621330e-02  9.89175960e-03 -4.67570201e-02 -4.92968261e-02\n",
      " -4.66732346e-02  1.85119212e-02 -1.43037522e-02  6.36090413e-02\n",
      " -8.02951008e-02  4.05227318e-02  5.48408367e-02  3.39601152e-02\n",
      "  4.68740053e-02  1.36475982e-02  4.05861475e-02  7.92822987e-03\n",
      "  1.15503244e-01 -1.29782269e-02 -1.09132621e-02  6.49238974e-02\n",
      " -2.03779414e-02  8.29743743e-02  4.70612757e-02 -1.12386085e-01\n",
      " -9.97546688e-03  2.86502559e-02 -3.29626538e-02  2.92402133e-02\n",
      " -4.99883182e-02 -7.39852665e-03 -6.86610788e-02  1.54224792e-02\n",
      "  2.85393801e-02 -2.27623265e-02 -7.08786678e-03  3.46387953e-33\n",
      "  1.00535393e-01  2.76806373e-02 -1.18886143e-01 -8.31663143e-03\n",
      "  1.07564982e-02 -6.05089637e-03 -1.61541905e-02  3.68626676e-02\n",
      " -1.18147619e-01  4.28505652e-02  5.12588806e-02  6.13651834e-02\n",
      " -5.04990481e-03 -5.88548444e-02  6.36577047e-03  1.97541434e-03\n",
      "  2.10786238e-02  7.18242303e-02 -8.74571130e-02  8.98227021e-02\n",
      " -6.67698458e-02  1.64771117e-02  7.71869766e-03  9.64160189e-02\n",
      " -5.98530611e-03  5.06181382e-02  6.44315556e-02  3.34278569e-02\n",
      " -6.65147454e-02  9.39634629e-03 -7.17762038e-02  3.26221846e-02\n",
      " -7.18062669e-02 -1.49450703e-02 -4.94165905e-03  1.08190745e-01\n",
      "  6.42905757e-02 -1.12142310e-01 -9.12282616e-03  1.55273895e-03\n",
      " -4.76264469e-02 -4.24384624e-02  8.18791986e-02 -4.61249612e-02\n",
      "  8.97361198e-04  4.67688451e-03 -5.52123189e-02  5.15677594e-02\n",
      " -6.42085671e-02  4.09936309e-02 -1.98340360e-02 -4.91785258e-02\n",
      "  3.59498663e-03  6.98357895e-02  1.28050921e-02 -3.30358259e-02\n",
      " -7.03806281e-02 -1.01913773e-02 -9.44083259e-02 -7.51283020e-03\n",
      " -2.38794982e-02 -4.30043004e-02 -1.29529750e-02  4.43171896e-02\n",
      " -4.71529588e-02 -2.26276238e-02 -1.46336826e-02 -5.31254848e-03\n",
      " -8.27344717e-04  9.84813739e-03  7.35706463e-02 -8.46427381e-02\n",
      " -1.14061467e-01 -3.55942883e-02  1.27100227e-02  8.47452581e-02\n",
      "  9.02831368e-03  2.08310112e-02 -3.39811556e-02  6.41425326e-02\n",
      " -1.07603386e-01 -4.64681722e-03 -2.02018991e-02  1.55828157e-02\n",
      " -6.57836348e-03 -7.85251707e-02  3.50788012e-02 -1.12334192e-01\n",
      " -3.91105600e-02  1.68346036e-02 -6.47353902e-02  3.74320187e-02\n",
      " -8.54043141e-02  4.98977751e-02 -5.51252328e-02 -1.79529351e-08\n",
      "  1.41176563e-02 -5.55693656e-02 -5.90728000e-02  1.61093250e-02\n",
      "  4.29887176e-02  9.07805003e-03  2.90159490e-02 -5.87360188e-02\n",
      "  1.58786215e-02  2.33781599e-02  4.86648716e-02  5.88877723e-02\n",
      "  4.62116562e-02  5.86222410e-02  4.05923575e-02 -7.72006437e-02\n",
      "  7.16023904e-04  8.97730328e-03  1.62806679e-02  8.13908223e-03\n",
      "  5.18525355e-02 -2.77281925e-02  2.90346630e-02  9.48402435e-02\n",
      "  4.33413796e-02  2.91407164e-02 -1.77315474e-02  2.13291626e-02\n",
      "  4.64902371e-02  6.35977983e-02  2.04788540e-02  4.79425378e-02\n",
      " -2.62871645e-02 -4.55872938e-02  8.79515521e-03 -4.78482768e-02\n",
      "  4.82628271e-02  6.38534352e-02  5.13028959e-03  1.05022015e-02\n",
      " -3.58512476e-02  3.71478684e-02 -9.78209265e-03  3.44478860e-02\n",
      "  1.92765780e-02 -2.69217882e-02  2.89505459e-02 -8.02743062e-03\n",
      " -5.82102314e-02 -2.38546245e-02  4.19977047e-02 -6.73853746e-03\n",
      "  6.23688661e-02  7.25138327e-03 -1.01980917e-01 -3.10983211e-02\n",
      " -4.43058424e-02 -1.68037489e-02 -4.00306806e-02 -1.69989262e-02\n",
      " -5.50214499e-02  2.15862528e-03 -8.71012732e-02  4.02802266e-02]\n"
     ]
    }
   ],
   "source": [
    "## Code here\n",
    "\n",
    "from sentence_transformers import SentenceTransformer\n",
    "\n",
    "embedding_model = SentenceTransformer('all-MiniLM-L6-v2')\n",
    "\n",
    "text = 'Today is a sunny day.'\n",
    "result = embedding_model.encode(text)\n",
    "\n",
    "print(f'Original text: {text}')\n",
    "print(f'Vector length: {len(result)}')\n",
    "print(f'Embedding vector: {result}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4576aea4-9471-49e5-997c-7427d0e5f9c5",
   "metadata": {},
   "source": [
    "4. Eventually make a custom function named `st_embed()` that you can reuse, where it receive input text as a parameters and return the embedding. Convert the result to a list using `tolist()` function, this will be useful when used for preparing the MongoDB document."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e7fe55c1-3f02-44dd-a426-f0daed03f655",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-5.77207468e-02,  2.38114316e-02, -4.45517071e-04,  8.69354308e-02,\n",
       "        4.62214313e-02,  3.00016571e-02,  8.20653215e-02,  6.98352326e-03,\n",
       "        1.36114284e-02, -2.21099500e-02, -5.80056608e-02, -3.94593813e-02,\n",
       "        3.89880687e-02,  5.68154827e-02,  1.16812531e-02, -2.18262542e-02,\n",
       "        1.38524882e-02, -4.20515314e-02, -2.16452852e-02, -4.42779697e-02,\n",
       "       -1.06043527e-02,  1.08371042e-01, -4.88939062e-02,  2.00086948e-03,\n",
       "        3.89887057e-02,  4.43795323e-02,  5.04380278e-03, -1.52779408e-02,\n",
       "       -5.70711493e-02, -1.05001044e-03, -3.28431316e-02,  3.55217829e-02,\n",
       "        2.01757737e-02, -2.72000227e-02, -1.45785976e-02, -7.56126409e-03,\n",
       "       -1.22766867e-02, -3.43601108e-02, -4.34343442e-02,  3.11367698e-02,\n",
       "        3.63502987e-02, -1.80783253e-02,  3.79497744e-02, -1.18687879e-02,\n",
       "        3.74216475e-02,  9.84410048e-02, -5.72689697e-02,  9.83748510e-02,\n",
       "        7.06411749e-02, -2.04079095e-02, -7.95387402e-02,  2.16492135e-02,\n",
       "       -1.01804703e-01,  3.87321375e-02, -8.83880816e-03,  7.01013803e-02,\n",
       "        2.75202319e-02, -1.42856002e-01, -9.34355147e-03, -1.47533668e-02,\n",
       "       -4.25591646e-03,  4.49905954e-02, -8.59382376e-02,  9.20855030e-02,\n",
       "        6.39724210e-02, -4.54948470e-02, -2.35664495e-03, -5.60578480e-02,\n",
       "        7.95160457e-02, -3.23930457e-02,  3.88103910e-03,  6.30002841e-02,\n",
       "        4.81518358e-03, -1.49411727e-02,  8.09388794e-03, -5.42680398e-02,\n",
       "       -7.35697759e-05, -1.24643736e-01,  4.14816998e-02,  8.87646899e-02,\n",
       "        1.54323680e-02, -6.67951107e-02,  3.10902819e-02,  7.44607970e-02,\n",
       "       -1.33044831e-02,  1.91070288e-02,  8.17659274e-02, -2.35531498e-02,\n",
       "       -2.92021018e-02, -1.33168017e-02, -9.14907381e-02, -3.70243080e-02,\n",
       "       -1.66934319e-02, -4.15767869e-03, -8.25869069e-02,  8.52850750e-02,\n",
       "       -4.95133363e-03,  1.08466372e-02, -1.12167872e-01,  1.20393112e-01,\n",
       "        1.06146550e-02, -6.07748553e-02,  5.08029852e-03,  5.64261563e-02,\n",
       "        3.81716527e-02, -5.32727316e-02,  3.09469402e-02, -8.42988938e-02,\n",
       "       -9.95934103e-03, -4.37554456e-02, -2.39035692e-02, -1.82877127e-02,\n",
       "       -1.41664103e-01,  1.21965930e-02,  5.94301708e-02, -1.23215532e-02,\n",
       "       -4.54699323e-02, -7.47904181e-03, -4.77409698e-02,  2.76044570e-02,\n",
       "       -8.57894402e-03,  4.60385494e-02, -1.12190535e-02,  4.01141047e-02,\n",
       "        2.99241617e-02, -9.95178372e-02,  6.64269850e-02, -1.01328664e-32,\n",
       "        7.69674480e-02,  8.68498930e-04,  1.03625305e-01,  4.35292572e-02,\n",
       "        9.30561498e-02, -3.83918323e-02, -4.36083488e-02,  3.49332243e-02,\n",
       "       -4.78165952e-04,  1.31439455e-02,  1.47862965e-02, -5.42654656e-02,\n",
       "       -4.64044102e-02, -4.45966944e-02,  7.64203742e-02, -9.67848822e-02,\n",
       "       -4.30707484e-02,  4.85612042e-02, -1.29129533e-02,  9.42735374e-02,\n",
       "       -1.58179216e-02,  1.94651354e-02,  4.23111692e-02,  1.59587078e-02,\n",
       "       -4.56693247e-02, -4.39842157e-02, -2.95784902e-02, -4.16270341e-04,\n",
       "        1.04406318e-02,  4.86364916e-05, -8.58077034e-03, -1.94921214e-02,\n",
       "        2.20074807e-03,  1.30491825e-02,  8.98448750e-03, -8.68272260e-02,\n",
       "        7.98391737e-03, -1.66016426e-02,  2.27183700e-02, -3.29986736e-02,\n",
       "        3.31051238e-02, -4.18378413e-02,  8.39027204e-03, -1.74776893e-02,\n",
       "       -1.49477657e-03,  3.24352160e-02,  8.89038015e-03, -2.47429982e-02,\n",
       "       -4.88413386e-02,  1.44413598e-02,  4.65729162e-02,  2.46317359e-03,\n",
       "        4.47877608e-02,  1.74186826e-02,  4.03912105e-02,  2.13339968e-04,\n",
       "        7.00267032e-02,  6.03650175e-02,  4.62916568e-02,  4.18139715e-03,\n",
       "       -5.42283170e-02, -3.14383805e-02, -3.17156352e-02, -4.52326052e-02,\n",
       "        2.70658382e-03,  4.22268808e-02, -3.45619619e-02,  4.31128480e-02,\n",
       "        1.71532054e-02, -7.86645990e-03,  4.05413210e-02,  5.93440048e-03,\n",
       "        7.70377815e-02,  3.02390270e-02,  2.57065310e-03, -9.96808242e-03,\n",
       "        4.34890427e-02,  2.96257646e-03,  4.11784239e-02,  8.53967667e-03,\n",
       "       -9.36251655e-02, -4.77219410e-02,  1.18294559e-01,  1.73173025e-02,\n",
       "       -3.96632142e-02,  4.32558544e-03, -3.40690576e-02,  7.43625127e-03,\n",
       "       -1.60919130e-02,  7.57594854e-02, -7.18518123e-02, -3.53825949e-02,\n",
       "        3.21484427e-03,  8.04052800e-02, -6.21404173e-03,  6.10814599e-33,\n",
       "        6.63840920e-02, -1.99745484e-02,  6.91655604e-03,  3.17427851e-02,\n",
       "        9.65054147e-03,  2.66177654e-02, -1.01368930e-02,  3.45622413e-02,\n",
       "       -5.37184626e-02,  6.33693784e-02,  8.72438680e-03,  7.54383905e-03,\n",
       "        8.42222124e-02, -1.09959524e-02,  1.15686551e-01,  1.16062522e-01,\n",
       "        8.38166103e-02,  9.80418921e-02,  3.94449197e-02,  5.66991828e-02,\n",
       "       -7.61459991e-02,  3.12827080e-02,  1.00039002e-02,  3.23552191e-02,\n",
       "       -3.24382037e-02,  5.31020090e-02,  5.50755560e-02,  4.08329815e-02,\n",
       "       -8.35969374e-02,  1.97913777e-02, -6.07338585e-02, -6.79332763e-02,\n",
       "        1.95486918e-02, -5.10625318e-02, -7.05736950e-02,  7.53289983e-02,\n",
       "       -2.50530150e-03, -8.10374618e-02, -5.91686480e-02, -4.48924974e-02,\n",
       "        7.69392848e-02, -5.44870384e-02,  1.14630714e-01,  5.48404828e-02,\n",
       "        1.90044176e-02, -7.03210384e-02, -7.36773387e-02, -1.57717858e-02,\n",
       "        2.19070762e-02,  9.61208493e-02,  9.55252163e-03, -4.48652618e-02,\n",
       "       -7.55447224e-02,  1.31035242e-02, -1.88207682e-02, -7.00068427e-03,\n",
       "       -1.64159119e-01,  4.09048283e-03, -3.28085292e-03, -3.18036228e-02,\n",
       "       -2.65060051e-04, -3.85611914e-02, -4.60399278e-02, -6.28035218e-02,\n",
       "       -1.50362141e-02,  2.29179915e-02, -5.55070080e-02,  5.57556236e-03,\n",
       "        9.78229381e-03,  9.58336145e-03,  1.58862397e-01, -4.07085121e-02,\n",
       "       -5.62569089e-02, -1.15700506e-01,  4.22224067e-02, -1.51018621e-02,\n",
       "        1.90765224e-02,  1.25112962e-02,  3.74512114e-02, -6.56818552e-03,\n",
       "       -4.99757230e-02,  7.42952060e-03, -5.61162904e-02,  2.22586375e-02,\n",
       "        6.61913380e-02, -8.00936967e-02,  3.54517065e-02, -1.53268583e-03,\n",
       "       -2.09270678e-02, -3.25839990e-03, -3.53058241e-02,  3.44416164e-02,\n",
       "       -9.64595657e-03,  7.45432824e-02, -1.92556977e-02, -2.00727452e-08,\n",
       "        2.21335888e-02, -8.92421303e-05, -5.09894677e-02,  3.05464417e-02,\n",
       "        2.90601682e-02,  2.98478510e-02, -2.29068436e-02, -9.57813486e-02,\n",
       "        2.44534444e-02,  3.11887749e-02,  1.19836978e-03,  9.93789807e-02,\n",
       "        7.42848590e-02,  6.98326156e-02, -1.27623659e-02, -6.71311319e-02,\n",
       "       -7.12668672e-02, -1.74626224e-02, -2.74251588e-02, -6.54122010e-02,\n",
       "       -4.32427675e-02, -1.50738098e-02,  5.34023717e-02,  4.29854915e-02,\n",
       "        5.38503518e-03,  4.94372100e-02,  3.67882103e-02,  6.46858737e-02,\n",
       "        7.72664398e-02,  4.93009500e-02, -1.07651107e-01,  1.29866395e-02,\n",
       "       -1.13459295e-02, -3.65850865e-05,  3.60844880e-02, -7.33626559e-02,\n",
       "        9.21104178e-02,  4.64732014e-02,  2.36332193e-02, -1.06981546e-01,\n",
       "       -2.72094123e-02,  1.74083803e-02, -2.88365055e-02, -1.69833675e-02,\n",
       "       -2.85497922e-02, -2.26868670e-02, -1.69798788e-02,  3.06768119e-02,\n",
       "       -1.09218080e-02, -5.91270998e-03, -7.31815584e-03,  1.11724064e-03,\n",
       "        9.42967832e-02,  7.12018162e-02, -3.29299048e-02, -1.94975603e-02,\n",
       "       -9.27254632e-02, -2.51254230e-03, -2.85538565e-02,  1.00937551e-02,\n",
       "       -7.15826154e-02, -3.79563645e-02, -1.11171462e-01,  6.04939871e-02],\n",
       "      dtype=float32)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Code here\n",
    "\n",
    "def st_embed(text):\n",
    "    embedding_model = SentenceTransformer('all-MiniLM-L6-v2')\n",
    "    result = embedding_model.encode(text)\n",
    "    return result\n",
    "\n",
    "st_embed('This is cold.')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4703481-f110-42a1-a49e-9cfe0a87d646",
   "metadata": {},
   "source": [
    "Great! you have created an embedded sentence!\n",
    "<hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18b3d3db-0caf-4c18-8016-ef92b7077eaf",
   "metadata": {},
   "source": [
    "### P2: Using local LLMs from `ollama`\n",
    "#### Method 1: using `requests`\n",
    "Ollama is a platform to run open-source LLMs locally. When running Ollama, by default, there will be a localhost with port `11434` started and served as long as Ollama is active. We can use this served localhost to interact with the local LLM models in Ollama.\n",
    "\n",
    "Once install Ollama, go to any browser, and try put this in the address bar: `localhost:11434` if the browser is showing `Ollama is running` then good news! Ollama is running and works!\n",
    "\n",
    "Then we will use the `requests` library to send web requests to interact with LLMs in Ollama\n",
    "\n",
    "1. import the library using\n",
    "\n",
    "   `import requests`\n",
    "   \n",
    "2. Then define the API url for Ollama embedding and the model to use\n",
    "\n",
    "    `url = 'http://localhost:11434/api/embeddings`\n",
    "    \n",
    "3.  use the model of `nomic-embed-text`\n",
    "\n",
    "    *Please make sure the model is installed in Ollama, or else, open the terminal and install the model using the command of: `ollama pull nomic-embed-text`*\n",
    "\n",
    "4. To interact, we will use the `post()` function in `requests` and pass the url, model and the text in following format:\n",
    "\n",
    "    <pre>\n",
    "    requests.post(\n",
    "        url,\n",
    "        json = {\n",
    "            \"model\" : model,\n",
    "            \"prompt\" : text\n",
    "        }\n",
    "    )</pre>\n",
    "\n",
    "5. To retrieve the embedding, extract the JSON structure of the response and filter the `embedding` key. Example: `response.json()['embedding']`\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "24b89f59-af12-40a4-872b-7034f17bb5ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Code here\n",
    "\n",
    "import requests\n",
    "\n",
    "url = 'http://localhost:11434/api/embeddings'\n",
    "model = 'nomic-embed-text' \n",
    "\n",
    "response = requests.post(\n",
    "        url,\n",
    "        json = {\n",
    "            \"model\" : model,\n",
    "            \"prompt\" : text\n",
    "        }\n",
    "    )\n",
    "\n",
    "result = response.json()['embedding']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a953f08-3ecc-4604-8fa9-afffc71a2964",
   "metadata": {},
   "source": [
    "6. Eventually make a custom function named `ollama_embed()` that you can reuse, where it receive input text as a parameters and return the embedding."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "16dbea5b-ddf4-469a-9440-53103447b26b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Code here\n",
    "def ollama_embed(text):\n",
    "    url = 'http://localhost:11434/api/embeddings'\n",
    "    model = 'nomic-embed-text'\n",
    "    response = requests.post(url, json={\"model\":model, \"prompt\":text})\n",
    "    result = response.json()['embedding']\n",
    "    return result\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aea3944b-036c-4cc7-9c67-fa458f4add16",
   "metadata": {},
   "source": [
    "#### Method 2: using `langchain`\n",
    "LangChain is a comprehensive library for LLM integration in Python. Using LangChain, we can directly interact with Ollama and use the object-oriented way to interact with the open source LLMs in Ollama.\n",
    "\n",
    "[*Reference: https://python.langchain.com/docs/introduction/*]\n",
    "\n",
    "Please make sure install the following library:\n",
    "- `langchain-ollama`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4aa98fa3-fcdb-423a-8fa8-9e46b57b4538",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: langchain-ollama in c:\\users\\benlc\\miniforge3\\lib\\site-packages (0.3.3)\n",
      "Requirement already satisfied: ollama<1.0.0,>=0.4.8 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langchain-ollama) (0.5.1)\n",
      "Requirement already satisfied: langchain-core<1.0.0,>=0.3.60 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langchain-ollama) (0.3.65)\n",
      "Requirement already satisfied: langsmith<0.4,>=0.3.45 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langchain-core<1.0.0,>=0.3.60->langchain-ollama) (0.3.45)\n",
      "Requirement already satisfied: tenacity!=8.4.0,<10.0.0,>=8.1.0 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langchain-core<1.0.0,>=0.3.60->langchain-ollama) (9.1.2)\n",
      "Requirement already satisfied: jsonpatch<2.0,>=1.33 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langchain-core<1.0.0,>=0.3.60->langchain-ollama) (1.33)\n",
      "Requirement already satisfied: PyYAML>=5.3 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langchain-core<1.0.0,>=0.3.60->langchain-ollama) (6.0.2)\n",
      "Requirement already satisfied: packaging<25,>=23.2 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langchain-core<1.0.0,>=0.3.60->langchain-ollama) (24.1)\n",
      "Requirement already satisfied: typing-extensions>=4.7 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langchain-core<1.0.0,>=0.3.60->langchain-ollama) (4.12.2)\n",
      "Requirement already satisfied: pydantic>=2.7.4 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langchain-core<1.0.0,>=0.3.60->langchain-ollama) (2.11.3)\n",
      "Requirement already satisfied: httpx>=0.27 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from ollama<1.0.0,>=0.4.8->langchain-ollama) (0.27.2)\n",
      "Requirement already satisfied: anyio in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from httpx>=0.27->ollama<1.0.0,>=0.4.8->langchain-ollama) (4.6.2.post1)\n",
      "Requirement already satisfied: certifi in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from httpx>=0.27->ollama<1.0.0,>=0.4.8->langchain-ollama) (2024.8.30)\n",
      "Requirement already satisfied: httpcore==1.* in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from httpx>=0.27->ollama<1.0.0,>=0.4.8->langchain-ollama) (1.0.7)\n",
      "Requirement already satisfied: idna in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from httpx>=0.27->ollama<1.0.0,>=0.4.8->langchain-ollama) (3.10)\n",
      "Requirement already satisfied: sniffio in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from httpx>=0.27->ollama<1.0.0,>=0.4.8->langchain-ollama) (1.3.1)\n",
      "Requirement already satisfied: h11<0.15,>=0.13 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from httpcore==1.*->httpx>=0.27->ollama<1.0.0,>=0.4.8->langchain-ollama) (0.14.0)\n",
      "Requirement already satisfied: jsonpointer>=1.9 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from jsonpatch<2.0,>=1.33->langchain-core<1.0.0,>=0.3.60->langchain-ollama) (3.0.0)\n",
      "Requirement already satisfied: orjson<4.0.0,>=3.9.14 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langsmith<0.4,>=0.3.45->langchain-core<1.0.0,>=0.3.60->langchain-ollama) (3.10.18)\n",
      "Requirement already satisfied: requests<3,>=2 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langsmith<0.4,>=0.3.45->langchain-core<1.0.0,>=0.3.60->langchain-ollama) (2.32.3)\n",
      "Requirement already satisfied: requests-toolbelt<2.0.0,>=1.0.0 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langsmith<0.4,>=0.3.45->langchain-core<1.0.0,>=0.3.60->langchain-ollama) (1.0.0)\n",
      "Requirement already satisfied: zstandard<0.24.0,>=0.23.0 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langsmith<0.4,>=0.3.45->langchain-core<1.0.0,>=0.3.60->langchain-ollama) (0.23.0)\n",
      "Requirement already satisfied: annotated-types>=0.6.0 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from pydantic>=2.7.4->langchain-core<1.0.0,>=0.3.60->langchain-ollama) (0.7.0)\n",
      "Requirement already satisfied: pydantic-core==2.33.1 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from pydantic>=2.7.4->langchain-core<1.0.0,>=0.3.60->langchain-ollama) (2.33.1)\n",
      "Requirement already satisfied: typing-inspection>=0.4.0 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from pydantic>=2.7.4->langchain-core<1.0.0,>=0.3.60->langchain-ollama) (0.4.0)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from requests<3,>=2->langsmith<0.4,>=0.3.45->langchain-core<1.0.0,>=0.3.60->langchain-ollama) (3.3.2)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from requests<3,>=2->langsmith<0.4,>=0.3.45->langchain-core<1.0.0,>=0.3.60->langchain-ollama) (2.2.3)\n"
     ]
    }
   ],
   "source": [
    "## \n",
    "\n",
    "!pip install langchain-ollama\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87e9ee88-3d23-43db-9eb3-33e1551901cc",
   "metadata": {},
   "source": [
    "Then to interact with Ollama,\n",
    "1. import the library: `from langchain_ollama import OllamaEmbeddings`\n",
    "2. Create an `OllamaEmbeddings` object and define the model to be used.\n",
    "\n",
    "    Example: `OllamaEmbeddings(model=xxx)`\n",
    "3. use the `embed_query()` function in the `OllamaEmbeddings` object, and pass in the text, it will return the embedding directly.\n",
    "\n",
    "    *The length of the vector is depending on the embedding model used*\n",
    "4. Eventually make a custom function named `ollama_embed_lc()` that you can reuse, where it receive input text as a parameters and return the embedding."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "a8fc24e7-de21-4b16-beeb-1ce19baaf01c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Embedding vector for example_text:\n",
      "[0.045759533, 0.048638687, -0.15519412, -0.07119232, 0.0648026, -0.006465471, 0.04468354, -0.0060434523, 0.017531652, -0.023755439, -0.043661922, 0.023491938, 0.004667976, 0.049296096, -0.064267315, 0.023731759, 0.077339865, -0.104143724, -0.018971408, 0.04803753, -0.021584792, -0.012928062, -0.031158226, -0.016311025, 0.06997232, 0.0154662365, 0.0044312472, 0.0031901796, -0.06508122, -0.031204365, 0.055903103, 0.002314534, -0.0034225208, -0.047710445, -0.022882627, -0.056000713, -0.020222919, 0.0852439, -0.0032614095, -0.04707582, 0.01913203, 0.0095965015, -0.029933885, -0.007653282, 0.02193616, -0.032393187, 0.08501813, 0.018890873, 0.02683022, -0.08258654, 0.0007857107, -0.044959612, 0.021984795, -0.047106717, 0.09804634, -0.00891882, 0.012330126, -0.04701691, -0.0027929994, -0.015128974, 0.028743735, 0.041340705, -0.096452326, 0.07011951, 0.028670367, -0.0043691387, -0.035164267, 0.057154763, 0.027297853, 0.0037948925, 0.043571137, 0.018515568, 0.020065185, -0.022460483, -0.0074138357, -0.021769544, -0.039279137, 0.025027249, 0.011338745, 0.027364548, 0.027977165, 0.033299327, 0.050663225, 0.016912347, 0.09786595, -0.014533487, -0.020908251, -0.05487543, -0.05524592, 0.038826663, 0.01670399, -0.009884299, 0.009894554, 0.039035533, -0.024182767, -0.020580905, -0.019282516, 0.049400788, 0.01293081, 0.0064355317, -0.060522463, -0.026796605, -0.05319647, -0.048619114, 0.0075755515, 0.046764743, 0.022290733, 0.018649379, -0.023108447, -0.033300016, -0.0044888817, 0.06781704, -0.056690183, -0.015292857, -0.033896334, 0.0033804434, 0.025298143, -0.027231224, -0.006386284, 0.02030717, 0.044910412, 0.023838831, 0.00046060368, 0.03315017, -0.02874207, 0.034514025, -0.06514486, 0.027070079, -0.002904258, -0.031004898, 0.0034127499, -0.018337695, -0.0051031113, -0.037511677, -0.00848703, 0.04212836, 0.004092627, -0.027006187, -0.024895405, 0.03448801, -0.0058552204, 0.0038701827, 0.020944947, -0.00606812, -0.017296935, -0.03240847, 0.009615463, -0.012264907, 0.017785186, 0.025629533, 0.011383167, 0.0066839587, 0.035239045, 0.008635627, 0.043789104, -0.0052285413, 0.031364907, 0.017212303, 0.053659804, 0.06524927, 0.05948098, 0.02671717, -0.051791713, 0.005370683, 0.006918157, -0.0408397, -0.024907833, 0.08798371, 0.03774144, -0.02306231, -0.028616551, -0.01991706, -0.027897794, 0.002274836, -0.034385163, -0.006619469, 0.015492204, -0.07761608, 0.057892613, -0.036065258, 0.031899266, -0.024343388, 0.009279975, -0.009658201, -0.04577932, -0.04940377, 0.027133683, -0.056841616, -0.030326486, -0.016023414, -0.048085086, 0.013437265, -0.045782857, -0.01862971, -0.04978105, -0.039795265, 0.083534695, 0.0026291842, 0.015585073, 0.0040132757, -0.016122226, 0.00355811, -0.016456278, 0.018662743, -0.02867133, -0.010145279, -0.023046633, -0.01615139, -0.049630724, 0.034060493, 0.06924005, -0.006197322, -0.015188285, 0.026188469, 0.05849834, -0.0010275929, 0.025547618, 0.00023491392, -0.061824135, -0.00095798, 0.0089295935, -0.004251213, 0.0015577201, -0.054784637, 0.04282353, 0.008664551, -0.042185314, -0.019685553, -0.018238403, 0.0013905959, -0.0021648055, -0.06393672, 0.03453737, 0.010225214, 0.04037904, 0.024816042, -0.026636373, 0.039156094, -0.023837041, -0.0032245037, -0.016164614, 0.038900334, 0.00035733194, -0.010860505, 0.011714603, -0.00082374737, 0.039380513, -0.07443746, 0.029152447, 0.05637623, -0.012046344, 0.027327381, 0.014282981, -0.008368469, 0.041101456, -0.057594344, -0.03537898, -0.0154033415, 0.0022943122, -0.0015034315, 0.032231413, -0.018238302, -0.015878374, -0.01779761, -0.11172402, -0.04859013, -0.03818233, -0.024494663, 0.040348966, 0.01757416, -0.033668403, 0.022227937, 0.02337419, 0.030976381, 0.019206982, -0.034719285, 0.009968288, -0.023026679, -0.0067266948, 0.032753102, -0.028697599, 0.020747028, -0.046041865, 0.032504153, -0.016248792, -0.0036401413, 0.016939305, -0.026286526, 0.023545595, 0.021684833, 0.013667047, 0.003005218, -0.017224394, 0.02830377, -0.019195052, 0.04127513, 0.062690586, -0.019194057, 0.031743586, 0.016919069, 0.0101932455, 0.048780855, 0.07342657, 0.037664432, 0.008705659, -0.001517378, 0.073122874, -0.04037562, 0.05208335, -0.007403942, -0.079533905, -0.020160673, -0.0033641693, -0.021660356, -0.073061995, 0.039181106, 0.02987938, 0.02124837, 0.030907327, -0.0694974, 0.023205606, -0.034491196, 0.035491142, 0.009212004, -0.025756052, 0.033990502, 0.015725395, -0.0059925546, -0.03534919, 0.009862472, -0.0060058264, 0.08025991, 0.013505, -0.029825866, 0.009940087, -0.006537473, -0.0063759545, -0.008045484, -0.035694562, 0.008354715, 0.050254256, -0.017746553, 0.026661312, -0.020539293, 0.06182979, -0.033861484, 0.0051168026, -5.9535407e-05, 0.025068983, 0.04177031, -0.027785353, -0.0036332274, 0.010643765, -0.041664116, -0.003886034, 0.019979283, 0.06557922, 0.007217694, 0.014121446, -0.0065129716, 0.089997075, -0.015793866, 0.021558072, -0.05495763, 0.013413434, 0.01576657, 0.026074272, 0.03998502, 0.03348541, 0.05265775, -0.018896908, -0.02928504, -0.010274808, 0.07294267, 0.03528429, 0.023991097, -0.086304076, -0.007677653, -0.010529411, -0.01865906, 0.009913347, 0.026850874, -0.019740276, -0.013952365, 0.019047907, -0.0037857275, 0.028676623, -0.005032616, 0.03735843, 0.023009432, -0.02027706, 0.038011067, -0.06307406, 0.033833135, 0.011083382, -0.03391525, 0.021671178, -0.0011157923, -0.024820833, 0.0033636268, -0.033216417, -0.028017448, -0.012660799, 0.011787705, -0.021823466, -0.020837579, -0.0034352343, -0.02230895, 0.015089891, 0.016136857, 0.028924203, 0.040100347, 0.008817413, -0.037581943, -0.010165776, -0.0027893754, 0.047282163, 0.022365946, -0.040692344, 0.029046394, -0.031968784, 0.09731696, -0.024147417, 0.019966159, 0.01156247, 0.0063185114, 0.0052434118, 0.052007217, 0.036298424, -0.056601685, -0.04916282, 0.03177803, 0.00033880462, 0.0073096324, 0.022090346, 0.00021776509, -0.02036579, 0.016405003, 0.035950683, 0.05706519, -0.017338429, -0.03920123, -0.02406978, -0.00071517675, 0.016142959, 0.10827309, 0.042992726, -0.036989093, -0.025758104, 0.0035161015, 0.0136671355, 0.041969642, 0.015790798, 0.021673799, 0.049419805, -0.018344712, 0.030864246, -0.02381913, -0.022799408, 0.03327233, 0.021970265, 0.031366218, 0.010454782, 0.024818202, 0.010220145, -0.01335, -0.0088200485, -0.0037847161, 0.02430311, 0.05948959, -0.051997706, 0.013977442, 0.038471688, -0.022575451, 0.050825763, 0.019862875, -0.03444498, -0.006705492, -0.008892798, 0.042839274, 0.02036083, -0.039968055, -0.029595628, -0.04772385, 0.03256378, 0.023345055, -0.0028227344, 0.036401495, 0.03686598, -0.005927293, 0.03915625, -0.0069759493, -0.013217706, 0.014612323, 0.0020731676, -0.040188942, -0.03163564, -0.0050948467, 0.041685175, -0.022323437, 0.011279073, 0.061341915, -0.0051512527, 0.00079530297, 0.0062378542, -0.03727613, -0.0012145117, -0.066485025, -0.075497106, 0.012535748, -0.014557583, 0.01880227, 0.014246309, -0.0019615965, 0.051280137, -0.03170694, -0.052896466, 0.036823984, -0.0671384, 0.04115877, 0.009565168, -0.0069058524, -0.016353775, -0.0045883204, -0.02208164, 0.011903849, 0.004511547, -0.06644236, 0.0835895, 0.019889574, 0.039071307, 0.049326807, -0.048128854, -0.035917822, 0.015146395, -0.04447779, -0.055832535, 0.028804231, 0.014708687, 0.013805331, -0.03141972, -0.0075513427, -0.021648347, 0.024478024, 0.022846092, 0.023346774, -0.04355045, -0.0034930022, 0.03451429, -0.04363747, -0.0002998105, -0.06625424, -0.06270869, -0.055128563, 0.057267908, -0.016347876, -0.037003543, -0.050665013, 0.01853889, -0.035030067, 0.026239825, -0.012350542, 0.019664764, 0.03231059, 0.013774986, -0.052369364, -0.012863875, 0.011564812, -0.004331657, 0.003068976, 0.03485091, -0.06343471, 0.045425687, -0.001855155, -0.04885167, -0.028509432, 0.018230012, 0.044274252, -0.058182675, -0.014092947, 0.019106323, -0.04425675, 0.014323717, 0.03788303, 0.0088630095, 0.02040445, -0.017420754, -0.047927257, -0.004185768, -0.033771787, -0.05377515, -0.035687298, -0.007510728, 0.009259628, 0.027580956, 0.0036836984, 0.008989755, -0.08042467, -0.012305786, -0.0182861, -0.008783691, 0.013888283, 0.047204748, -0.0612652, 0.025924148, 0.057780664, 0.052995577, 0.0011402303, 0.024987245, -0.05170307, 0.011936166, 0.018020663, -0.05098344, -0.000117568474, 0.028840393, -0.02329945, 0.0672548, -0.033265408, 0.0009784875, -0.007170137, -0.009674431, -0.0342388, 0.056305844, 0.0024608145, 0.01401894, 0.020248963, -0.0592408, 0.011849685, 0.00496248, 0.045739934, -0.021781862, 0.016816907, -0.09115252, -0.02377472, -0.05335381, 0.022224935, 0.024732655, 0.01281346, 0.018150255, 0.056615386, -0.015473143, 0.0008947039, -0.05130108, 0.006771626, 0.015928816, 0.045276567, 0.032403715, 0.035879586, 0.038192663, -0.0040001115, 0.06910797, 0.081756584, 0.0064481255, -0.037693188, 0.039534297, 0.046582084, 0.02361754, 0.011017611, -0.042885683, -0.0035148931, -0.011982139, 0.0003303471, -0.0324202, -0.0385958, 0.03757731, 0.006850329, -0.0029817456, -0.06001258, -0.014333176, -0.04178847, 0.048465125, -0.019835087, 0.02176953, -0.0291099, -0.034973703, -0.0015024504, 0.054909904, 0.0664665, -0.011309187, -0.051899474, -0.030648505, 0.035101324, 0.07150444, 0.0030989307, 0.021508206, -0.091112874, 0.03775538, -0.027488379, -0.037137628, -0.027662965, -0.021342985, -0.007280225, 0.0010494575, -0.013939916, 0.015429602, -0.007118276, 0.043868992, -0.0126407575, 0.022692785, 0.038678247, -0.10212335, 0.03662172, 0.03459112, 0.03176954, -0.064880416, -0.016027063, -0.0071560727, -0.031072786, -0.023929039, -0.08066451, 0.022979354, 0.010984128, -0.004953401, 0.022655465, -0.012359828, 0.047687415, -0.039514318, -0.043817002, -0.0011769449, 0.00050088426, 0.031590454, 0.012195396, -0.028290186, -0.029137224, -0.025689224, -0.06304915, 0.04936825, -0.021954497, 0.042822715, -0.05803463, -0.004179277, -0.006060949, -0.0075853267, 0.03682848, -0.022140374, 0.0027400302, -0.053756144, -0.015900878, -0.020816507, 0.021157466, -0.04951482, 0.013408796, -0.040575124, -0.006058936, 0.033441532, -0.007807047, -0.035117526, 0.018414153, 0.021584027, -0.038635433, 0.026258754, -0.07360143, -0.018051008, -0.007856532, -0.008476739, -0.021793688, -0.029268159, 0.05833559, 0.13534002, 0.003488636, 0.0051489985, -0.0046799094, -0.02070443, -0.016072929, -0.00554511, -0.03710534, -0.0027015263, -0.03463833]\n",
      "\n",
      "Embedding vector for another_text:\n",
      "[0.026192248, 0.008458271, -0.19767463, 0.0032073206, -0.006914501, -0.06215369, -0.0067519834, -0.043576144, 0.0018684756, 0.010811652, -0.031634364, 0.08860652, 0.048885353, -0.026488965, 0.017679526, 0.004158747, 0.020226382, -0.03291721, 0.05304678, 0.0035064055, -0.040177558, -0.06345004, -0.07519308, -0.006136123, 0.13586015, 0.015731992, 0.031936683, 0.06254314, -0.02115675, 0.055933442, 0.044198666, -0.050201025, -0.0031256604, -0.0128628295, -0.008507192, 0.009109181, 0.030145273, -0.02673228, 7.867648e-06, 0.033383023, 0.037706275, -0.057532087, 0.011214999, 0.012549935, 0.009421218, -0.012643784, 0.004091649, -0.024007393, 0.009756794, 0.0010901074, -0.021411026, -0.0016115573, -0.011031708, -0.06133637, 0.029313603, 0.015340541, 0.016657291, 0.0019695226, -0.00832094, -0.018175429, 0.07814791, 0.0044880915, -0.08504668, 0.0734417, -0.0027980537, -0.028598657, -0.03558929, 0.044836156, -0.038039256, -0.03812104, 0.006221626, -0.004487914, 0.036364455, 0.021737393, -0.037318226, 0.0134177385, -0.012350291, -0.034313124, 0.0015423972, -0.00997495, -0.0144702755, -0.022682581, 0.111088835, 0.0040593427, 0.037568595, -0.01651291, -0.0077735847, -0.018987512, -0.088957734, 0.030502766, 0.017577438, 0.00048206197, 0.022149341, 0.009622311, -0.051581316, 0.0066720624, -0.00093683624, 0.06292225, -0.017434128, -0.014468671, 0.044795144, 0.018692955, -0.024744367, 0.034971382, 0.028271245, -0.04015592, -0.021990929, -0.04240855, -0.008778622, -0.028574714, -0.0020749243, 0.0069318404, -0.01649615, -0.059199832, 0.016374912, -0.05167382, 0.023839595, -0.01958037, -0.010950807, 0.00496954, -0.038439363, -0.022418221, 0.0037691437, -0.019516133, 0.022685084, 0.024374822, -0.0034406865, 0.026626365, -0.023772651, -0.052985527, -0.019325651, -0.027747605, 0.0037756057, -0.0018051175, 0.007657281, 0.03463133, -0.03278553, -0.036492642, -0.009266286, 0.07255249, -0.0048762267, 0.00090862444, 0.0051374407, -0.010249345, -0.034317352, -0.03319235, 0.030498771, -0.006533836, -0.025705751, 0.03304626, -0.030352086, -0.0028522834, 0.03318344, 0.033751223, 0.017728947, -0.04402303, -0.013372754, 0.013441779, -0.007684707, 0.02262546, 0.015848309, 0.04282815, -0.038195517, 0.024592951, -0.029423833, -0.05323609, -0.050352793, 0.021200268, 0.01414908, 0.0837358, -0.08982013, -0.013766831, 0.0025798725, -0.0352714, 0.042994186, -0.014280384, -0.0292885, -0.0021514425, 0.018961508, 0.023868669, 0.061054964, 0.029965704, -0.011443421, 0.030954335, -0.023706883, -0.026004804, 0.029222475, 0.00346391, -0.049488317, 0.03153417, -0.024821164, 0.0953712, -0.075862706, -0.0064055645, -0.021815425, -0.03171628, 0.0014571386, 0.05335393, 0.0048360657, -0.0021445756, -0.006981674, -0.05874594, -0.05546804, -0.01921854, -0.05639555, 0.021225633, 0.008661672, 0.02409745, 0.038094953, 0.068696916, 0.031002602, -0.06138292, -0.007570622, -0.015648015, 0.014037202, -0.023864374, 0.0028855528, -0.043407366, 0.009056058, 0.048243158, 0.05165709, -0.0032417825, 0.034217045, -0.022669382, 0.042629138, -0.01365826, -0.035115264, 0.007886318, 0.0040203687, 0.007963572, 0.006493625, -0.0032858704, 0.025466718, 0.014428251, -0.040104087, -0.003349566, -0.008363521, 0.06668499, -0.02870399, -0.014649198, -0.016484825, 0.043365847, 0.033877186, -0.0037832612, -0.055483058, 0.0683696, -0.08125651, -0.060504828, -0.024814129, 0.02655159, 0.016749103, 0.0087536415, 0.0244664, 0.025544433, -0.0057299524, -0.044935167, -0.054374408, -0.020685969, 0.00577904, -0.048601564, 0.0091132885, 0.00018301267, 0.024291186, -0.06539325, -0.012672563, 0.0065370006, -0.035972014, -0.009333127, 0.011729827, -0.082215145, 0.04476343, 0.034834642, 0.023154546, 0.009858586, 0.0008491074, 0.047272548, 0.0011065609, -0.041033115, 0.016301932, -0.0073238034, 0.00090040884, -0.03562619, 0.0024176554, -0.015021269, -0.04688377, 0.05630616, 0.058423273, -0.020981919, 0.029598804, 0.027555853, 0.01235136, -0.031738713, 0.0010066474, 0.09678743, -0.023944264, -0.011367374, 0.07499113, -0.033623207, 0.017478894, -0.008521238, 0.03826433, 0.0044977497, 0.029636724, -0.013081245, -0.010983799, -0.04344481, 0.0023447552, -0.028571755, 0.07183927, 0.027360873, 0.029121395, -0.017196806, -0.016945755, 0.000113925846, -0.040505707, -0.026751023, 0.00083533075, 0.0014510376, 0.06548681, -0.0024489015, 0.030371206, -0.043750465, 0.008181516, -0.065371156, 0.008046134, -0.019268302, 0.0071685407, 0.07885992, -0.027173964, 0.01775268, 0.01587458, 0.042046037, 0.011878248, 0.00018139134, 0.0028006139, 0.020635713, -0.001784521, -0.0231372, -0.029274834, 0.05602995, -0.01706871, 0.00025464233, 0.024667874, 0.023950988, 0.032681394, -0.015295711, 0.0100553315, -0.060290314, 0.017682567, 0.041677877, -0.052667227, 0.023821624, 0.020809555, -0.009937655, 0.026650796, -0.037187226, 0.0625968, 0.027680747, 0.04100464, -0.016058221, 0.0412214, 0.022591237, 0.015107627, -0.013656071, 0.024564264, -0.010627482, -0.0048602615, -0.010518647, 0.025615571, 0.017591286, -0.02016867, -0.0148430895, -0.03328761, 0.053728595, -0.020462945, 0.016444072, -0.06843878, -0.043761276, -0.0050630863, 0.0091878455, 0.03481541, -0.075009435, 0.03034358, 0.007395828, 0.02332166, 0.056265745, 0.031117823, -0.004428985, 0.057405323, -0.052523706, 0.0014100478, 0.0129924845, -0.040506374, -0.019693144, -0.010879034, -0.013946723, 0.05245103, -0.04176443, 0.0016995417, 0.015816152, -0.043122977, -0.051157814, 0.04352082, -0.05206209, 0.021411547, 0.00035737187, -0.016401732, -0.021233097, 0.07411114, -0.025951885, 0.012514897, 0.044896774, 0.0036810497, -0.06170595, 0.00053923955, 0.028526133, 0.0048291436, -0.0192516, -0.03174805, 0.007611753, 0.09305553, 0.04873626, -0.00964516, -0.014799121, 0.010596808, 0.015750425, 0.006245849, -0.0027847586, 0.040379934, -0.08324597, 0.032964822, -0.010255911, -0.01383543, -0.04217563, -0.039911747, 0.0074398783, -0.016823437, 0.033878192, 0.06598271, 0.0043906188, 0.037779395, -0.015132987, 0.007623623, 0.0034791843, -0.011340385, -0.0002768671, 0.12696247, -0.04348944, -0.0013276689, -0.007236607, 0.005735423, -0.019641131, 0.037336722, 0.047118854, 0.07571528, -0.05669702, 0.02879071, 0.0059175617, 0.023800872, 0.0114858, -0.00037206613, 0.05631551, -0.053356756, 0.02576524, 0.07564778, -0.05401272, -0.0051672594, 0.0026814407, -0.003204657, 0.056526333, 0.009909822, -0.01652192, 0.061326515, -0.021636862, -0.023164421, -0.016915854, -0.010097837, 0.014162129, -0.0054566683, 0.063840695, 0.026314858, -0.057899702, -0.04428649, -0.07343734, 0.013769649, 0.06008664, 0.0019373484, -0.02048132, 0.0115013225, 0.022433901, -0.00826909, 0.0012542745, 0.0047503486, -0.027328415, -0.0043309233, -0.01326593, -0.009023458, 0.0076488596, 0.046721082, -0.035255052, -0.035302483, 0.019788735, 0.0023219741, 0.019005965, 0.035645712, -0.023908721, -0.0007636444, -0.02314386, -0.026812488, 0.022657687, -0.032755055, 0.018820237, 0.034826558, -0.038977996, 0.009876152, -0.07944868, -0.0016630263, 0.03607594, 0.009504544, -0.005980089, 0.004341766, -0.07211569, -0.05250179, 0.026610991, -0.03736011, 0.021940233, 0.019463329, -0.086058445, 0.038393952, -0.00904726, -0.0054524173, -0.024682026, -0.03253506, -0.028443007, -0.060922503, -0.0026961726, -0.01560723, 0.046555653, 0.028086782, 0.0042626183, 0.0038889647, 0.024484482, -0.00397988, -0.0046060057, -0.027209058, 0.048106372, -0.07968846, 0.02559711, 0.024431322, -0.04764882, 0.060227405, -0.010027473, 0.017799767, 0.015036802, -0.00420619, -0.002211307, -0.021975521, -0.037336294, -0.018949812, -0.0020807688, -0.011204264, 0.019001873, 0.036652125, -0.005991159, -0.0024518417, -0.06409791, -0.00679778, 0.065044105, 0.0125877615, -0.011032001, 0.048334267, -0.054931004, -0.028931104, 0.05374979, 0.024388699, -0.011221624, 0.033333164, 0.026140066, -0.037084673, -0.021564202, 0.008241474, -0.06645808, -0.0011943924, 0.07261693, 0.06994108, 0.056639608, -0.0071823737, -0.057612967, -0.0018902516, -0.062543385, -0.023229895, 0.020222638, 0.00753277, -0.0056741624, 0.011430447, -0.009867179, -0.008654537, -0.027700385, -0.008051256, -0.03648484, -0.011319378, 0.014108965, -0.012246336, -0.0052290135, -0.036154363, 0.027934924, 0.010485639, -0.015962673, -0.043709885, 0.039142188, 0.030853676, 0.010529148, -0.016784666, -0.034601845, -0.004221457, 0.03781478, 0.022055518, -0.005445782, 0.0008689567, -0.016535437, -0.014581489, -0.042501975, 0.01176763, -0.044001564, -0.0036557226, -0.027879072, -0.04638926, -0.03167176, 0.0046511316, 0.04090046, 0.006889683, 0.007934664, -0.04912404, -0.03462984, -0.08228393, -0.015349876, 0.026112124, -0.011171141, 0.021875069, 0.045998488, 0.027816286, 0.03019834, -0.043721545, 0.05100154, 0.090439186, -0.022832423, -0.023797147, 0.03896583, 0.0312666, 0.00178914, 0.07642609, 0.072205685, 0.039205417, -0.06567367, 0.00524488, -0.023044148, 0.005548172, -0.054504998, -0.038819678, -0.013334211, -0.0045401524, 0.034543604, 0.010150646, 0.029664919, 0.07946539, 0.01574748, 0.00029284714, 0.011089505, -0.036819216, -0.0033061993, -0.0039892965, 0.04190209, -0.02849639, 0.0061746216, -0.039062776, 0.03846477, 0.04229639, 0.029215489, -0.02768431, -0.006767986, 0.01817356, -0.06300412, 0.00377739, 0.020185996, 0.0052173208, -0.04356586, 0.046170194, -0.017003648, 0.0046003903, -0.06772388, 0.05064031, -0.06536275, 0.0009998763, -0.028614411, -0.023344697, 0.017536003, -0.027111612, 0.006004942, -0.056905963, 0.0039491192, 0.001022651, 0.03239564, 0.00813658, 0.084233, -0.078945756, 0.05084728, -0.023463951, 0.015361294, 0.0342703, 0.01256276, -0.03316585, 0.014882477, -0.03167938, 0.040986933, 0.07026153, 0.04056717, 0.0076429225, -0.045936733, -0.024320377, 0.02335988, 0.03053883, -0.029649017, -0.061183907, 0.0035408246, -0.023960454, 0.0038072316, 0.0124286385, -0.042188477, 0.0071156616, -0.057601977, 0.024873922, -0.010602854, -0.035681643, 0.044427443, 0.05351226, -0.009264101, -0.022387585, -0.025532108, 0.0037944664, -0.047229894, -0.04621606, 0.030263325, 0.0048377574, 0.03255931, -0.01227964, 0.053059414, -0.022447025, 0.024892462, -0.012757697, -0.0011957828, -0.0128085585, -0.021288993, -0.019542713, 0.02855199, 0.014571905, 0.07982109, 0.05076621, 0.039251085, 0.088719346, 0.008177279, 0.010118872, -0.03479769, 0.030740958, 0.061392795, -0.008493339, -0.0046278914, -0.052893795, -0.04934888]\n"
     ]
    }
   ],
   "source": [
    "## Code here\n",
    "from langchain_ollama import OllamaEmbeddings\n",
    "\n",
    "embedding = OllamaEmbeddings(model=\"nomic-embed-text\")\n",
    "\n",
    "example_text = \"This is a test sentence to embed.\"\n",
    "embedding_vector = embedding.embed_query(example_text)\n",
    "\n",
    "\n",
    "print(\"Embedding vector for example_text:\")\n",
    "print(embedding_vector)\n",
    "\n",
    "def ollama_embed_lc(input_text: str) -> list:\n",
    "    \"\"\"\n",
    "    Takes input text and returns its embedding vector using Ollama embeddings.\n",
    "    \"\"\"\n",
    "    return embedding.embed_query(input_text)\n",
    "\n",
    "\n",
    "another_text = \"LMAO!!! x9 ANIVIA MID!\"\n",
    "another_vector = ollama_embed_lc(another_text)\n",
    "\n",
    "print(\"\\nEmbedding vector for another_text:\")\n",
    "print(another_vector)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6920204a-3e3f-4fea-a061-46c4e30b7cc2",
   "metadata": {},
   "source": [
    "Great! Now you can embed text using local LLMs in Ollama.\n",
    "<hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5c11146-d4d4-4ac1-b8ce-43aef380dda3",
   "metadata": {},
   "source": [
    "## Embed protein data and store in MongoDB\n",
    "### 1. Preparing the data and embeddings\n",
    "Now we will perform an example use case that using protein data (retrieved from Uniprot database) and create a MongoDB database to store this data and the embedding of it for vector search.\n",
    "\n",
    "1. Read the data from the csv file of `uniprot_proteins.csv` Please download it from e-learning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "559f9853-9fa1-49c4-b133-1a34fff6fe79",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>entry_name</th>\n",
       "      <th>protein_name</th>\n",
       "      <th>gene_names</th>\n",
       "      <th>organism</th>\n",
       "      <th>sequence_length</th>\n",
       "      <th>sequence</th>\n",
       "      <th>function</th>\n",
       "      <th>subcellular_location</th>\n",
       "      <th>pathways</th>\n",
       "      <th>domains</th>\n",
       "      <th>go_molecular_function</th>\n",
       "      <th>go_biological_process</th>\n",
       "      <th>go_cellular_component</th>\n",
       "      <th>keywords</th>\n",
       "      <th>ec_numbers</th>\n",
       "      <th>protein_families</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P12821</td>\n",
       "      <td>ACE_HUMAN</td>\n",
       "      <td>Angiotensin-converting enzyme</td>\n",
       "      <td>['ACE']</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1306</td>\n",
       "      <td>MGAASGRRGPGLLLPLPLLLLLPPQPALALDPGLQPGNFSADEAGA...</td>\n",
       "      <td>Isoform produced by alternative promoter usage...</td>\n",
       "      <td>Cell membrane; Secreted</td>\n",
       "      <td>[]</td>\n",
       "      <td>['Peptidase M2 1', 'Peptidase M2 2']</td>\n",
       "      <td>['actin binding', 'bradykinin receptor binding...</td>\n",
       "      <td>['amyloid-beta metabolic process', 'angiogenes...</td>\n",
       "      <td>['basal plasma membrane', 'brush border membra...</td>\n",
       "      <td>['3D-structure', 'Alternative promoter usage',...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P23368</td>\n",
       "      <td>MAOM_HUMAN</td>\n",
       "      <td>NAD-dependent malic enzyme, mitochondrial</td>\n",
       "      <td>['ME2']</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>584</td>\n",
       "      <td>MLSRLRVVSTTCTLACRHLHIKEKGKPLMLNPRTNKGMAFTLQERQ...</td>\n",
       "      <td>NAD-dependent mitochondrial malic enzyme that ...</td>\n",
       "      <td>Mitochondrion matrix</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>['electron transfer activity', 'malate dehydro...</td>\n",
       "      <td>['malate metabolic process', 'pyruvate metabol...</td>\n",
       "      <td>['intracellular membrane-bounded organelle', '...</td>\n",
       "      <td>['3D-structure', 'Acetylation', 'Allosteric en...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P49427</td>\n",
       "      <td>UB2R1_HUMAN</td>\n",
       "      <td>Ubiquitin-conjugating enzyme E2 R1</td>\n",
       "      <td>['CDC34']</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>236</td>\n",
       "      <td>MARPLVPSSQKALLLELKGLQEEPVEGFRVTLVDEGDLYNWEVAIF...</td>\n",
       "      <td>E2 ubiquitin-conjugating enzyme that accepts u...</td>\n",
       "      <td>Cytoplasm; Nucleus</td>\n",
       "      <td>['Protein modification; protein ubiquitination']</td>\n",
       "      <td>['UBC core']</td>\n",
       "      <td>['ATP binding', 'ubiquitin conjugating enzyme ...</td>\n",
       "      <td>['cellular response to interferon-beta', 'DNA ...</td>\n",
       "      <td>['cytosol', 'nuclear speck', 'nucleoplasm', 'n...</td>\n",
       "      <td>['3D-structure', 'ATP-binding', 'Cell cycle', ...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P62256</td>\n",
       "      <td>UBE2H_HUMAN</td>\n",
       "      <td>Ubiquitin-conjugating enzyme E2 H</td>\n",
       "      <td>['UBE2H']</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>183</td>\n",
       "      <td>MSSPSPGKRRMDTDVVKLIESKHEVTILGGLNEFVVKFYGPQGTPY...</td>\n",
       "      <td>Accepts ubiquitin from the E1 complex and cata...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Protein modification; protein ubiquitination']</td>\n",
       "      <td>['UBC core']</td>\n",
       "      <td>['ATP binding', 'ubiquitin conjugating enzyme ...</td>\n",
       "      <td>['proteasome-mediated ubiquitin-dependent prot...</td>\n",
       "      <td>['cytosol', 'nucleus']</td>\n",
       "      <td>['3D-structure', 'Acetylation', 'Alternative s...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P61077</td>\n",
       "      <td>UB2D3_HUMAN</td>\n",
       "      <td>Ubiquitin-conjugating enzyme E2 D3</td>\n",
       "      <td>['UBE2D3']</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>147</td>\n",
       "      <td>MALKRINKELSDLARDPPAQCSAGPVGDDMFHWQATIMGPNDSPYQ...</td>\n",
       "      <td>Accepts ubiquitin from the E1 complex and cata...</td>\n",
       "      <td>Cell membrane; Endosome membrane</td>\n",
       "      <td>['Protein modification; protein ubiquitination']</td>\n",
       "      <td>['UBC core']</td>\n",
       "      <td>['ATP binding', 'ubiquitin conjugating enzyme ...</td>\n",
       "      <td>['apoptotic process', 'DNA repair', 'negative ...</td>\n",
       "      <td>['cytosol', 'endosome membrane', 'extracellula...</td>\n",
       "      <td>['3D-structure', 'Alternative splicing', 'Apop...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  uniprot_id   entry_name                               protein_name  \\\n",
       "0     P12821    ACE_HUMAN              Angiotensin-converting enzyme   \n",
       "1     P23368   MAOM_HUMAN  NAD-dependent malic enzyme, mitochondrial   \n",
       "2     P49427  UB2R1_HUMAN         Ubiquitin-conjugating enzyme E2 R1   \n",
       "3     P62256  UBE2H_HUMAN          Ubiquitin-conjugating enzyme E2 H   \n",
       "4     P61077  UB2D3_HUMAN         Ubiquitin-conjugating enzyme E2 D3   \n",
       "\n",
       "   gene_names      organism  sequence_length  \\\n",
       "0     ['ACE']  Homo sapiens             1306   \n",
       "1     ['ME2']  Homo sapiens              584   \n",
       "2   ['CDC34']  Homo sapiens              236   \n",
       "3   ['UBE2H']  Homo sapiens              183   \n",
       "4  ['UBE2D3']  Homo sapiens              147   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MGAASGRRGPGLLLPLPLLLLLPPQPALALDPGLQPGNFSADEAGA...   \n",
       "1  MLSRLRVVSTTCTLACRHLHIKEKGKPLMLNPRTNKGMAFTLQERQ...   \n",
       "2  MARPLVPSSQKALLLELKGLQEEPVEGFRVTLVDEGDLYNWEVAIF...   \n",
       "3  MSSPSPGKRRMDTDVVKLIESKHEVTILGGLNEFVVKFYGPQGTPY...   \n",
       "4  MALKRINKELSDLARDPPAQCSAGPVGDDMFHWQATIMGPNDSPYQ...   \n",
       "\n",
       "                                            function  \\\n",
       "0  Isoform produced by alternative promoter usage...   \n",
       "1  NAD-dependent mitochondrial malic enzyme that ...   \n",
       "2  E2 ubiquitin-conjugating enzyme that accepts u...   \n",
       "3  Accepts ubiquitin from the E1 complex and cata...   \n",
       "4  Accepts ubiquitin from the E1 complex and cata...   \n",
       "\n",
       "               subcellular_location  \\\n",
       "0           Cell membrane; Secreted   \n",
       "1              Mitochondrion matrix   \n",
       "2                Cytoplasm; Nucleus   \n",
       "3                               NaN   \n",
       "4  Cell membrane; Endosome membrane   \n",
       "\n",
       "                                           pathways  \\\n",
       "0                                                []   \n",
       "1                                                []   \n",
       "2  ['Protein modification; protein ubiquitination']   \n",
       "3  ['Protein modification; protein ubiquitination']   \n",
       "4  ['Protein modification; protein ubiquitination']   \n",
       "\n",
       "                                domains  \\\n",
       "0  ['Peptidase M2 1', 'Peptidase M2 2']   \n",
       "1                                    []   \n",
       "2                          ['UBC core']   \n",
       "3                          ['UBC core']   \n",
       "4                          ['UBC core']   \n",
       "\n",
       "                               go_molecular_function  \\\n",
       "0  ['actin binding', 'bradykinin receptor binding...   \n",
       "1  ['electron transfer activity', 'malate dehydro...   \n",
       "2  ['ATP binding', 'ubiquitin conjugating enzyme ...   \n",
       "3  ['ATP binding', 'ubiquitin conjugating enzyme ...   \n",
       "4  ['ATP binding', 'ubiquitin conjugating enzyme ...   \n",
       "\n",
       "                               go_biological_process  \\\n",
       "0  ['amyloid-beta metabolic process', 'angiogenes...   \n",
       "1  ['malate metabolic process', 'pyruvate metabol...   \n",
       "2  ['cellular response to interferon-beta', 'DNA ...   \n",
       "3  ['proteasome-mediated ubiquitin-dependent prot...   \n",
       "4  ['apoptotic process', 'DNA repair', 'negative ...   \n",
       "\n",
       "                               go_cellular_component  \\\n",
       "0  ['basal plasma membrane', 'brush border membra...   \n",
       "1  ['intracellular membrane-bounded organelle', '...   \n",
       "2  ['cytosol', 'nuclear speck', 'nucleoplasm', 'n...   \n",
       "3                             ['cytosol', 'nucleus']   \n",
       "4  ['cytosol', 'endosome membrane', 'extracellula...   \n",
       "\n",
       "                                            keywords ec_numbers  \\\n",
       "0  ['3D-structure', 'Alternative promoter usage',...         []   \n",
       "1  ['3D-structure', 'Acetylation', 'Allosteric en...         []   \n",
       "2  ['3D-structure', 'ATP-binding', 'Cell cycle', ...         []   \n",
       "3  ['3D-structure', 'Acetylation', 'Alternative s...         []   \n",
       "4  ['3D-structure', 'Alternative splicing', 'Apop...         []   \n",
       "\n",
       "  protein_families  \n",
       "0               []  \n",
       "1               []  \n",
       "2               []  \n",
       "3               []  \n",
       "4               []  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Code here\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "proteins_data = pd.read_csv(\"uniprot_proteins.csv\")\n",
    "proteins_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9a25194-2eee-4a79-9330-d2fe57ec9457",
   "metadata": {},
   "source": [
    "2. Decide the protein info for embedding. There are several fields and we will decide to use the fields of `protein_name` and `function` to be combined as a \"sentence\" that prepared to be embedding.\n",
    "\n",
    "    Let say the format of the text for every row would be: <pre>Proteins: protein_name | Function: function</pre>\n",
    "\n",
    "   Convert this into a function named `format_searchable()` that takes a row of the table as input. So that can be re-use later on all the rows in the dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "e15f6610-4064-420f-a845-1b8ba7886e2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Code here\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "#df\n",
    "def format_searchable(row):\n",
    "    searchable_text = f'{row['protein_name']} | {row['function']}'\n",
    "    return searchable_text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "5c329553",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Angiotensin-converting enzyme | Isoform produced by alternative promoter usage that is specifically expressed in spermatocytes and adult testis, and which is required for male fertility (PubMed:1651327, PubMed:1668266). In contrast to somatic isoforms, only contains one catalytic domain (PubMed:1651327, PubMed:1668266). Acts as a dipeptidyl carboxypeptidase that removes dipeptides from the C-terminus of substrates (PubMed:1668266, PubMed:24297181). The identity of substrates that are needed for male fertility is unknown (By similarity). May also have a glycosidase activity which releases GPI-anchored proteins from the membrane by cleaving the mannose linkage in the GPI moiety. The GPIase activity was reported to be essential for the egg-binding ability of the sperm (By similarity). This activity is however unclear and has been challenged by other groups, suggesting that it may be indirect (By similarity)'"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "format_searchable(proteins_data.iloc[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3cfa03b-f9b2-4cc5-af94-da2a3dc8766b",
   "metadata": {},
   "source": [
    "3. Then, loop every data entry on the `protein_data` imported at Step 1. Create a document that stores the data in following format:\n",
    "<pre>{\n",
    "        'uniprot_id':\n",
    "        'entry_name':\n",
    "        'protein_name':\n",
    "        'gene_names':\n",
    "        'organism':\n",
    "        'sequence_length':\n",
    "        'sequence':\n",
    "        'function':\n",
    "        'subcellular_location':\n",
    "        'domains':\n",
    "        'keywords':\n",
    "        'go_molecular_function':\n",
    "        'go_biological_process':\n",
    "        'go_cellular_component':\n",
    "        'searchable_text': #this is the text produced in Step 2\n",
    "        'embedding': #embedding result from st_embed() using searchable_text\n",
    "   }</pre>\n",
    "\n",
    "   #### Using the function `st_embed()` from above\n",
    "4. During the processing, create the embedding for the `searchable_text` as well using the `st_embed()` function defined above.\n",
    "\n",
    "5. This process may need additional time to complete\n",
    "   \n",
    "6. Create a variable `protein_docs` at the end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "d67dce70-84bf-4db0-bf39-b78a150f1066",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing document: 349/349\n",
      "Completed\n"
     ]
    }
   ],
   "source": [
    "## Code here\n",
    "\n",
    "from IPython.display import clear_output\n",
    "\n",
    "protein_docs = []\n",
    "\n",
    "for idx, row in proteins_data.iterrows():\n",
    "    clear_output(wait=True)\n",
    "    print(f'Processing document: {idx+1}/{proteins_data.shape[0]}')\n",
    "    searchable_text = format_searchable(row)\n",
    "\n",
    "    doc = {\n",
    "        'uniprot_id': row['uniprot_id'],\n",
    "        'entry_name': row['entry_name'],\n",
    "        'protein_name': row['protein_name'],\n",
    "        'gene_names': row['gene_names'],\n",
    "        'organism': row['organism'],\n",
    "        'sequence_length': int(row['sequence_length']),\n",
    "        'sequence': row['sequence'],\n",
    "        'function': row['function'],\n",
    "        'subcellular_location': row['subcellular_location'],\n",
    "        'domains': row['domains'],\n",
    "        'keywords': row['keywords'],\n",
    "        'go_molecular_function': row['go_molecular_function'],\n",
    "        'go_biological_process': row['go_biological_process'],\n",
    "        'go_cellular_component': row['go_cellular_component'],\n",
    "        'searchable_text': searchable_text,\n",
    "        'embedding': ollama_embed(searchable_text)\n",
    "    }\n",
    "\n",
    "    protein_docs.append(doc)\n",
    "\n",
    "print(\"Completed\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a12c043-9485-4c3b-8945-247b54ca869f",
   "metadata": {},
   "source": [
    "### 2. Ingest into MongoDB\n",
    "#### Initialize connection with MongoDB Atlas\n",
    "1. Establish the connection to MongoDB Atlas. First create a new account in MongoDB Atlas and create a free tier MongoDB database. Then obtain the URI and establish the connection with `pymongo`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "ccf28356-71f1-40ca-8a60-5a0e6b0d16b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully connected to MongoDB Atlas!\n"
     ]
    }
   ],
   "source": [
    "## Code here\n",
    "\n",
    "from pymongo import MongoClient\n",
    "\n",
    "MONGO_URI = \"mongodb+srv://benLab4:benlab4ass@cluster0.goopkpm.mongodb.net/\"\n",
    "\n",
    "try:\n",
    "    client = MongoClient(MONGO_URI)\n",
    "\n",
    "    client.admin.command(\"ping\")\n",
    "\n",
    "    db = client[\"imported\"]\n",
    "    collection = db[\"uniprot_proteins\"]\n",
    "\n",
    "    print(\"Successfully connected to MongoDB Atlas!\")\n",
    "\n",
    "except Exception as e:\n",
    "    print(\"Failed to connect to MongoDB Atlas:\")\n",
    "    print(e)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a08a47b-84a5-4886-a165-6ec5c183add7",
   "metadata": {},
   "source": [
    "#### Insert documents\n",
    "2. Then insert all the documents into the MongoDB Atlas database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "65369b31-9e6b-4ade-8458-6c24a9a9e094",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Inserted batch 1.0/7.96\n",
      "Inserted batch 2.0/7.96\n",
      "Inserted batch 3.0/7.96\n",
      "Inserted batch 4.0/7.96\n",
      "Inserted batch 5.0/7.96\n",
      "Inserted batch 6.0/7.96\n",
      "Inserted batch 7.0/7.96\n"
     ]
    }
   ],
   "source": [
    "## Code here\n",
    "batch_size = 50\n",
    "\n",
    "for i in range(0, len(protein_docs), batch_size):\n",
    "    batch = protein_docs[i:i+batch_size]\n",
    "    collection.insert_many(batch)\n",
    "    print(f\"Inserted batch {i/batch_size + 1}/{(len(protein_docs)-1)/batch_size + 1}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cba8755e-2261-4712-bb20-346716191fe1",
   "metadata": {},
   "source": [
    "#### Creating vector index\n",
    "3. Then create a vector search index in the MongoDB Atlas. In this case we will use `langchain-mongodb` and `langchain-ollama` library\n",
    "\n",
    "    In here will use the `MongoDBAtlasVectorSearch` in the `langchain-mongodb` library. To use the function, requires the following information:\n",
    "\n",
    "   <pre>\n",
    "       MongoDBAtlasVectorSearch(\n",
    "           collection = \n",
    "           embedding = \n",
    "           index_name = \n",
    "           relevance_score_fn = \n",
    "       )\n",
    "   </pre>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "69954222",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting langchain\n",
      "  Downloading langchain-0.3.26-py3-none-any.whl.metadata (7.8 kB)\n",
      "Collecting langchain-mongodb\n",
      "  Downloading langchain_mongodb-0.6.2-py3-none-any.whl.metadata (1.7 kB)\n",
      "Requirement already satisfied: langchain-ollama in c:\\users\\benlc\\miniforge3\\lib\\site-packages (0.3.3)\n",
      "Requirement already satisfied: pymongo in c:\\users\\benlc\\miniforge3\\lib\\site-packages (4.12.0)\n",
      "Collecting langchain-core<1.0.0,>=0.3.66 (from langchain)\n",
      "  Downloading langchain_core-0.3.66-py3-none-any.whl.metadata (5.8 kB)\n",
      "Collecting langchain-text-splitters<1.0.0,>=0.3.8 (from langchain)\n",
      "  Downloading langchain_text_splitters-0.3.8-py3-none-any.whl.metadata (1.9 kB)\n",
      "Requirement already satisfied: langsmith>=0.1.17 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langchain) (0.3.45)\n",
      "Requirement already satisfied: pydantic<3.0.0,>=2.7.4 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langchain) (2.11.3)\n",
      "Collecting SQLAlchemy<3,>=1.4 (from langchain)\n",
      "  Downloading sqlalchemy-2.0.41-cp312-cp312-win_amd64.whl.metadata (9.8 kB)\n",
      "Requirement already satisfied: requests<3,>=2 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langchain) (2.32.3)\n",
      "Requirement already satisfied: PyYAML>=5.3 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langchain) (6.0.2)\n",
      "Collecting lark<2.0.0,>=1.1.9 (from langchain-mongodb)\n",
      "  Downloading lark-1.2.2-py3-none-any.whl.metadata (1.8 kB)\n",
      "Requirement already satisfied: numpy>=1.26 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langchain-mongodb) (2.0.2)\n",
      "Requirement already satisfied: ollama<1.0.0,>=0.4.8 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langchain-ollama) (0.5.1)\n",
      "Requirement already satisfied: dnspython<3.0.0,>=1.16.0 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from pymongo) (2.7.0)\n",
      "Requirement already satisfied: tenacity!=8.4.0,<10.0.0,>=8.1.0 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langchain-core<1.0.0,>=0.3.66->langchain) (9.1.2)\n",
      "Requirement already satisfied: jsonpatch<2.0,>=1.33 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langchain-core<1.0.0,>=0.3.66->langchain) (1.33)\n",
      "Requirement already satisfied: packaging<25,>=23.2 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langchain-core<1.0.0,>=0.3.66->langchain) (24.1)\n",
      "Requirement already satisfied: typing-extensions>=4.7 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langchain-core<1.0.0,>=0.3.66->langchain) (4.12.2)\n",
      "Requirement already satisfied: httpx<1,>=0.23.0 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langsmith>=0.1.17->langchain) (0.27.2)\n",
      "Requirement already satisfied: orjson<4.0.0,>=3.9.14 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langsmith>=0.1.17->langchain) (3.10.18)\n",
      "Requirement already satisfied: requests-toolbelt<2.0.0,>=1.0.0 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langsmith>=0.1.17->langchain) (1.0.0)\n",
      "Requirement already satisfied: zstandard<0.24.0,>=0.23.0 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from langsmith>=0.1.17->langchain) (0.23.0)\n",
      "Requirement already satisfied: annotated-types>=0.6.0 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from pydantic<3.0.0,>=2.7.4->langchain) (0.7.0)\n",
      "Requirement already satisfied: pydantic-core==2.33.1 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from pydantic<3.0.0,>=2.7.4->langchain) (2.33.1)\n",
      "Requirement already satisfied: typing-inspection>=0.4.0 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from pydantic<3.0.0,>=2.7.4->langchain) (0.4.0)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from requests<3,>=2->langchain) (3.3.2)\n",
      "Requirement already satisfied: idna<4,>=2.5 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from requests<3,>=2->langchain) (3.10)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from requests<3,>=2->langchain) (2.2.3)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from requests<3,>=2->langchain) (2024.8.30)\n",
      "Collecting greenlet>=1 (from SQLAlchemy<3,>=1.4->langchain)\n",
      "  Downloading greenlet-3.2.3-cp312-cp312-win_amd64.whl.metadata (4.2 kB)\n",
      "Requirement already satisfied: anyio in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from httpx<1,>=0.23.0->langsmith>=0.1.17->langchain) (4.6.2.post1)\n",
      "Requirement already satisfied: httpcore==1.* in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from httpx<1,>=0.23.0->langsmith>=0.1.17->langchain) (1.0.7)\n",
      "Requirement already satisfied: sniffio in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from httpx<1,>=0.23.0->langsmith>=0.1.17->langchain) (1.3.1)\n",
      "Requirement already satisfied: h11<0.15,>=0.13 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from httpcore==1.*->httpx<1,>=0.23.0->langsmith>=0.1.17->langchain) (0.14.0)\n",
      "Requirement already satisfied: jsonpointer>=1.9 in c:\\users\\benlc\\miniforge3\\lib\\site-packages (from jsonpatch<2.0,>=1.33->langchain-core<1.0.0,>=0.3.66->langchain) (3.0.0)\n",
      "Downloading langchain-0.3.26-py3-none-any.whl (1.0 MB)\n",
      "   ---------------------------------------- 0.0/1.0 MB ? eta -:--:--\n",
      "   -------------------- ------------------- 0.5/1.0 MB 2.8 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 1.0/1.0 MB 2.7 MB/s eta 0:00:00\n",
      "Downloading langchain_mongodb-0.6.2-py3-none-any.whl (59 kB)\n",
      "Downloading langchain_core-0.3.66-py3-none-any.whl (438 kB)\n",
      "Downloading langchain_text_splitters-0.3.8-py3-none-any.whl (32 kB)\n",
      "Downloading lark-1.2.2-py3-none-any.whl (111 kB)\n",
      "Downloading sqlalchemy-2.0.41-cp312-cp312-win_amd64.whl (2.1 MB)\n",
      "   ---------------------------------------- 0.0/2.1 MB ? eta -:--:--\n",
      "   --------- ------------------------------ 0.5/2.1 MB 2.1 MB/s eta 0:00:01\n",
      "   -------------- ------------------------- 0.8/2.1 MB 2.2 MB/s eta 0:00:01\n",
      "   ------------------- -------------------- 1.0/2.1 MB 1.9 MB/s eta 0:00:01\n",
      "   ------------------- -------------------- 1.0/2.1 MB 1.9 MB/s eta 0:00:01\n",
      "   ------------------- -------------------- 1.0/2.1 MB 1.9 MB/s eta 0:00:01\n",
      "   ------------------------ --------------- 1.3/2.1 MB 972.7 kB/s eta 0:00:01\n",
      "   ----------------------------- ---------- 1.6/2.1 MB 1.1 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 2.1/2.1 MB 1.3 MB/s eta 0:00:00\n",
      "Downloading greenlet-3.2.3-cp312-cp312-win_amd64.whl (297 kB)\n",
      "Installing collected packages: lark, greenlet, SQLAlchemy, langchain-core, langchain-text-splitters, langchain, langchain-mongodb\n",
      "  Attempting uninstall: langchain-core\n",
      "    Found existing installation: langchain-core 0.3.65\n",
      "    Uninstalling langchain-core-0.3.65:\n",
      "      Successfully uninstalled langchain-core-0.3.65\n",
      "Successfully installed SQLAlchemy-2.0.41 greenlet-3.2.3 langchain-0.3.26 langchain-core-0.3.66 langchain-mongodb-0.6.2 langchain-text-splitters-0.3.8 lark-1.2.2\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%pip install langchain langchain-mongodb langchain-ollama pymongo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "b9e539ce-0a50-4058-8eee-dcdf766ec429",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Code here\n",
    "\n",
    "from langchain_mongodb import MongoDBAtlasVectorSearch\n",
    "from langchain_ollama import OllamaEmbeddings\n",
    "\n",
    "embed_model = OllamaEmbeddings(model='nomic-embed-text')\n",
    "\n",
    "vectorstore = MongoDBAtlasVectorSearch(\n",
    "    collection = collection,\n",
    "    embedding = embed_model,\n",
    "    index_name = \"prot_index\",\n",
    "    relevance_score_fn='cosine'\n",
    ")\n",
    "\n",
    "vectorstore.create_vector_search_index(dimensions=768)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50cc18bd-df46-4835-acaf-267a6ee1d191",
   "metadata": {},
   "source": [
    "Great! you have successfully embed the protein data and stored into MongoDB Atlas\n",
    "<hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c64e3dac-f6e8-4689-90aa-e9771cb86d43",
   "metadata": {},
   "source": [
    "## Vector Search in MongoDB Atlas\n",
    "### Testing the vector search\n",
    "1. Create a vector search pipeline using the `aggregate` in MongoDB and use the pipeline to perform the search.\n",
    "\n",
    "   The pipeline for vector search will use the `$vectorSearch` operator. The pipeline can be generally view as follows for the vector search stage:\n",
    "\n",
    "   <pre>\n",
    "       {\n",
    "            \"$vectorSearch\": {\n",
    "                \"index\": \"prot_index\",\n",
    "                \"path\": \"embedding\",\n",
    "                \"queryVector\": query_embedding,\n",
    "                \"numCandidates\": min(limit * 10, 1000),\n",
    "                \"limit\": \n",
    "            }\n",
    "        }\n",
    "   </pre>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "21070a12-9927-487f-8835-a7ae1d5accf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Code here\n",
    "\n",
    "query = \"enzyme that catalyzes protein breakdown\"\n",
    "query_embedding = ollama_embed(query)\n",
    "limit = 10\n",
    "\n",
    "pipeline = [\n",
    "    {\n",
    "        \"$vectorSearch\": {\n",
    "            \"index\": \"prot_index\",\n",
    "            \"path\": \"embedding\",\n",
    "            \"queryVector\": query_embedding,\n",
    "            \"numCandidates\": min(limit * 10, 1000),\n",
    "            \"limit\": limit\n",
    "        }\n",
    "    },\n",
    "    {\n",
    "        \"$project\": {\n",
    "            \"uniprot_id\": 1,\n",
    "            \"protein_name\": 1,\n",
    "            \"organism\": 1,\n",
    "            \"function\": 1,\n",
    "            \"keywords\": 1,\n",
    "            \"searchable_text\": 1,\n",
    "            \"score\": {\"$meta\": \"vectorSearchScore\"}  # Include similarity score\n",
    "        }\n",
    "    }\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "4df7fd43",
   "metadata": {},
   "outputs": [],
   "source": [
    "result = list(collection.aggregate(pipeline))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c14b0b6-5af3-418f-a138-7283a430859a",
   "metadata": {},
   "source": [
    "2. Create a function called `search_vector()` that accepts:\n",
    "    - query in text,\n",
    "    - the limit of results\n",
    "    - MongoDB collection object\n",
    "   \n",
    "   <br>as the input, and return the search result from MongoDB."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "84178804-6dba-41d7-9930-25cc7f9f2343",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Code here\n",
    "def search_vector(query, limit, collection):\n",
    "    query_embedding = ollama_embed(query)\n",
    "    pipeline = [\n",
    "        {\n",
    "            \"$vectorSearch\": {\n",
    "                \"index\": \"prot_index\",\n",
    "                \"path\": \"embedding\",\n",
    "                \"queryVector\": query_embedding,\n",
    "                \"numCandidates\": min(limit * 10, 1000),\n",
    "                \"limit\": limit\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            \"$project\": {\n",
    "                \"uniprot_id\": 1,\n",
    "                \"protein_name\": 1,\n",
    "                \"organism\": 1,\n",
    "                \"function\": 1,\n",
    "                \"keywords\": 1,\n",
    "                \"searchable_text\": 1,\n",
    "                \"score\": {\"$meta\": \"vectorSearchScore\"}  # Include similarity score\n",
    "            }\n",
    "        }\n",
    "    ]\n",
    "    result = list(collection.aggregate(pipeline))\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "933c0b05",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'_id': ObjectId('685ebb252516af4a76508da7'),\n",
       "  'uniprot_id': 'P17275',\n",
       "  'protein_name': 'Transcription factor JunB',\n",
       "  'organism': 'Homo sapiens',\n",
       "  'function': \"Transcription factor involved in regulating gene activity following the primary growth factor response. Binds to the DNA sequence 5'-TGA[GC]TCA-3'. Heterodimerizes with proteins of the FOS family to form an AP-1 transcription complex, thereby enhancing its DNA binding activity to an AP-1 consensus sequence and its transcriptional activity (By similarity)\",\n",
       "  'keywords': \"['Acetylation', 'DNA-binding', 'Isopeptide bond', 'Nucleus', 'Phosphoprotein', 'Proteomics identification', 'Reference proteome', 'Transcription', 'Transcription regulation', 'Ubl conjugation']\",\n",
       "  'searchable_text': \"Proteins: Transcription factor JunB | Function: Transcription factor involved in regulating gene activity following the primary growth factor response. Binds to the DNA sequence 5'-TGA[GC]TCA-3'. Heterodimerizes with proteins of the FOS family to form an AP-1 transcription complex, thereby enhancing its DNA binding activity to an AP-1 consensus sequence and its transcriptional activity (By similarity)\",\n",
       "  'score': 0.8881714940071106},\n",
       " {'_id': ObjectId('685ec5a12516af4a76508f04'),\n",
       "  'uniprot_id': 'P17275',\n",
       "  'protein_name': 'Transcription factor JunB',\n",
       "  'organism': 'Homo sapiens',\n",
       "  'function': \"Transcription factor involved in regulating gene activity following the primary growth factor response. Binds to the DNA sequence 5'-TGA[GC]TCA-3'. Heterodimerizes with proteins of the FOS family to form an AP-1 transcription complex, thereby enhancing its DNA binding activity to an AP-1 consensus sequence and its transcriptional activity (By similarity)\",\n",
       "  'keywords': \"['Acetylation', 'DNA-binding', 'Isopeptide bond', 'Nucleus', 'Phosphoprotein', 'Proteomics identification', 'Reference proteome', 'Transcription', 'Transcription regulation', 'Ubl conjugation']\",\n",
       "  'searchable_text': \"Transcription factor JunB | Transcription factor involved in regulating gene activity following the primary growth factor response. Binds to the DNA sequence 5'-TGA[GC]TCA-3'. Heterodimerizes with proteins of the FOS family to form an AP-1 transcription complex, thereby enhancing its DNA binding activity to an AP-1 consensus sequence and its transcriptional activity (By similarity)\",\n",
       "  'score': 0.8870155811309814},\n",
       " {'_id': ObjectId('685ebb252516af4a76508da8'),\n",
       "  'uniprot_id': 'P17535',\n",
       "  'protein_name': 'Transcription factor JunD',\n",
       "  'organism': 'Homo sapiens',\n",
       "  'function': \"Transcription factor binding AP-1 sites (PubMed:9989505). Heterodimerizes with proteins of the FOS family to form an AP-1 transcription factor complex, thereby enhancing their DNA binding activity to an AP-1 consensus sequence 3'-TGA[GC]TCA-5' and enhancing their transcriptional activity (PubMed:28981703, PubMed:9989505)\",\n",
       "  'keywords': \"['3D-structure', 'Activator', 'Disulfide bond', 'DNA-binding', 'Nucleus', 'Phosphoprotein', 'Proteomics identification', 'Reference proteome', 'Transcription', 'Transcription regulation']\",\n",
       "  'searchable_text': \"Proteins: Transcription factor JunD | Function: Transcription factor binding AP-1 sites (PubMed:9989505). Heterodimerizes with proteins of the FOS family to form an AP-1 transcription factor complex, thereby enhancing their DNA binding activity to an AP-1 consensus sequence 3'-TGA[GC]TCA-5' and enhancing their transcriptional activity (PubMed:28981703, PubMed:9989505)\",\n",
       "  'score': 0.8697136044502258},\n",
       " {'_id': ObjectId('685ebb252516af4a76508dab'),\n",
       "  'uniprot_id': 'Q00059',\n",
       "  'protein_name': 'Transcription factor A, mitochondrial',\n",
       "  'organism': 'Homo sapiens',\n",
       "  'function': 'Binds to the mitochondrial light strand promoter and functions in mitochondrial transcription regulation (PubMed:29445193, PubMed:32183942). Component of the mitochondrial transcription initiation complex, composed at least of TFB2M, TFAM and POLRMT that is required for basal transcription of mitochondrial DNA (PubMed:29149603). In this complex, TFAM recruits POLRMT to a specific promoter whereas TFB2M induces structural changes in POLRMT to enable promoter opening and trapping of the DNA non-template strand (PubMed:20410300). Required for accurate and efficient promoter recognition by the mitochondrial RNA polymerase (PubMed:22037172). Promotes transcription initiation from the HSP1 and the light strand promoter by binding immediately upstream of transcriptional start sites (PubMed:22037172). Is able to unwind DNA (PubMed:22037172). Bends the mitochondrial light strand promoter DNA into a U-turn shape via its HMG boxes (PubMed:1737790). Required for maintenance of normal levels of mitochondrial DNA (PubMed:19304746, PubMed:22841477). May play a role in organizing and compacting mitochondrial DNA (PubMed:22037171)',\n",
       "  'keywords': \"['3D-structure', 'Activator', 'Alternative splicing', 'Direct protein sequencing', 'Disease variant', 'DNA-binding', 'Mitochondrion', 'Mitochondrion nucleoid', 'Phosphoprotein', 'Primary mitochondrial disease', 'Proteomics identification', 'Reference proteome', 'Repeat', 'Transcription', 'Transcription regulation', 'Transit peptide']\",\n",
       "  'searchable_text': 'Proteins: Transcription factor A, mitochondrial | Function: Binds to the mitochondrial light strand promoter and functions in mitochondrial transcription regulation (PubMed:29445193, PubMed:32183942). Component of the mitochondrial transcription initiation complex, composed at least of TFB2M, TFAM and POLRMT that is required for basal transcription of mitochondrial DNA (PubMed:29149603). In this complex, TFAM recruits POLRMT to a specific promoter whereas TFB2M induces structural changes in POLRMT to enable promoter opening and trapping of the DNA non-template strand (PubMed:20410300). Required for accurate and efficient promoter recognition by the mitochondrial RNA polymerase (PubMed:22037172). Promotes transcription initiation from the HSP1 and the light strand promoter by binding immediately upstream of transcriptional start sites (PubMed:22037172). Is able to unwind DNA (PubMed:22037172). Bends the mitochondrial light strand promoter DNA into a U-turn shape via its HMG boxes (PubMed:1737790). Required for maintenance of normal levels of mitochondrial DNA (PubMed:19304746, PubMed:22841477). May play a role in organizing and compacting mitochondrial DNA (PubMed:22037171)',\n",
       "  'score': 0.8694534301757812},\n",
       " {'_id': ObjectId('685ec5a12516af4a76508f05'),\n",
       "  'uniprot_id': 'P17535',\n",
       "  'protein_name': 'Transcription factor JunD',\n",
       "  'organism': 'Homo sapiens',\n",
       "  'function': \"Transcription factor binding AP-1 sites (PubMed:9989505). Heterodimerizes with proteins of the FOS family to form an AP-1 transcription factor complex, thereby enhancing their DNA binding activity to an AP-1 consensus sequence 3'-TGA[GC]TCA-5' and enhancing their transcriptional activity (PubMed:28981703, PubMed:9989505)\",\n",
       "  'keywords': \"['3D-structure', 'Activator', 'Disulfide bond', 'DNA-binding', 'Nucleus', 'Phosphoprotein', 'Proteomics identification', 'Reference proteome', 'Transcription', 'Transcription regulation']\",\n",
       "  'searchable_text': \"Transcription factor JunD | Transcription factor binding AP-1 sites (PubMed:9989505). Heterodimerizes with proteins of the FOS family to form an AP-1 transcription factor complex, thereby enhancing their DNA binding activity to an AP-1 consensus sequence 3'-TGA[GC]TCA-5' and enhancing their transcriptional activity (PubMed:28981703, PubMed:9989505)\",\n",
       "  'score': 0.8620858788490295},\n",
       " {'_id': ObjectId('685ebb252516af4a76508d9e'),\n",
       "  'uniprot_id': 'O60675',\n",
       "  'protein_name': 'Transcription factor MafK',\n",
       "  'organism': 'Homo sapiens',\n",
       "  'function': 'Since they lack a putative transactivation domain, the small Mafs behave as transcriptional repressors when they dimerize among themselves (PubMed:9150357). However, they act as transcriptional activators by dimerizing with other (usually larger) basic-zipper proteins, such as NFE2, NFE2L1/NRF1, NFE2L2/NRF2 and NFE2L3/NRF3, and recruiting them to specific DNA-binding sites (PubMed:8932385, PubMed:9150357). Small Maf proteins heterodimerize with Fos and may act as competitive repressors of the NF-E2 transcription factor (PubMed:9150357)',\n",
       "  'keywords': \"['DNA-binding', 'Isopeptide bond', 'Nucleus', 'Phosphoprotein', 'Proteomics identification', 'Reference proteome', 'Repressor', 'Transcription', 'Transcription regulation', 'Ubl conjugation']\",\n",
       "  'searchable_text': 'Proteins: Transcription factor MafK | Function: Since they lack a putative transactivation domain, the small Mafs behave as transcriptional repressors when they dimerize among themselves (PubMed:9150357). However, they act as transcriptional activators by dimerizing with other (usually larger) basic-zipper proteins, such as NFE2, NFE2L1/NRF1, NFE2L2/NRF2 and NFE2L3/NRF3, and recruiting them to specific DNA-binding sites (PubMed:8932385, PubMed:9150357). Small Maf proteins heterodimerize with Fos and may act as competitive repressors of the NF-E2 transcription factor (PubMed:9150357)',\n",
       "  'score': 0.8607642650604248},\n",
       " {'_id': ObjectId('685ec5a12516af4a76508f08'),\n",
       "  'uniprot_id': 'Q00059',\n",
       "  'protein_name': 'Transcription factor A, mitochondrial',\n",
       "  'organism': 'Homo sapiens',\n",
       "  'function': 'Binds to the mitochondrial light strand promoter and functions in mitochondrial transcription regulation (PubMed:29445193, PubMed:32183942). Component of the mitochondrial transcription initiation complex, composed at least of TFB2M, TFAM and POLRMT that is required for basal transcription of mitochondrial DNA (PubMed:29149603). In this complex, TFAM recruits POLRMT to a specific promoter whereas TFB2M induces structural changes in POLRMT to enable promoter opening and trapping of the DNA non-template strand (PubMed:20410300). Required for accurate and efficient promoter recognition by the mitochondrial RNA polymerase (PubMed:22037172). Promotes transcription initiation from the HSP1 and the light strand promoter by binding immediately upstream of transcriptional start sites (PubMed:22037172). Is able to unwind DNA (PubMed:22037172). Bends the mitochondrial light strand promoter DNA into a U-turn shape via its HMG boxes (PubMed:1737790). Required for maintenance of normal levels of mitochondrial DNA (PubMed:19304746, PubMed:22841477). May play a role in organizing and compacting mitochondrial DNA (PubMed:22037171)',\n",
       "  'keywords': \"['3D-structure', 'Activator', 'Alternative splicing', 'Direct protein sequencing', 'Disease variant', 'DNA-binding', 'Mitochondrion', 'Mitochondrion nucleoid', 'Phosphoprotein', 'Primary mitochondrial disease', 'Proteomics identification', 'Reference proteome', 'Repeat', 'Transcription', 'Transcription regulation', 'Transit peptide']\",\n",
       "  'searchable_text': 'Transcription factor A, mitochondrial | Binds to the mitochondrial light strand promoter and functions in mitochondrial transcription regulation (PubMed:29445193, PubMed:32183942). Component of the mitochondrial transcription initiation complex, composed at least of TFB2M, TFAM and POLRMT that is required for basal transcription of mitochondrial DNA (PubMed:29149603). In this complex, TFAM recruits POLRMT to a specific promoter whereas TFB2M induces structural changes in POLRMT to enable promoter opening and trapping of the DNA non-template strand (PubMed:20410300). Required for accurate and efficient promoter recognition by the mitochondrial RNA polymerase (PubMed:22037172). Promotes transcription initiation from the HSP1 and the light strand promoter by binding immediately upstream of transcriptional start sites (PubMed:22037172). Is able to unwind DNA (PubMed:22037172). Bends the mitochondrial light strand promoter DNA into a U-turn shape via its HMG boxes (PubMed:1737790). Required for maintenance of normal levels of mitochondrial DNA (PubMed:19304746, PubMed:22841477). May play a role in organizing and compacting mitochondrial DNA (PubMed:22037171)',\n",
       "  'score': 0.8597841858863831},\n",
       " {'_id': ObjectId('685ebb252516af4a76508d9d'),\n",
       "  'uniprot_id': 'Q00403',\n",
       "  'protein_name': 'Transcription initiation factor IIB',\n",
       "  'organism': 'Homo sapiens',\n",
       "  'function': \"General transcription factor that plays a role in transcription initiation by RNA polymerase II (Pol II). Involved in the pre-initiation complex (PIC) formation and Pol II recruitment at promoter DNA (PubMed:12931194, PubMed:1517211, PubMed:1876184, PubMed:1946368, PubMed:27193682, PubMed:3029109, PubMed:3818643, PubMed:7601352, PubMed:8413225, PubMed:8515820, PubMed:8516311, PubMed:8516312, PubMed:9420329). Together with the TATA box-bound TBP forms the core initiation complex and provides a bridge between TBP and the Pol II-TFIIF complex (PubMed:8413225, PubMed:8504927, PubMed:8515820, PubMed:8516311, PubMed:8516312). Released from the PIC early following the onset of transcription during the initiation and elongation transition and reassociates with TBP during the next transcription cycle (PubMed:7601352). Associates with chromatin to core promoter-specific regions (PubMed:12931194, PubMed:24441171). Binds to two distinct DNA core promoter consensus sequence elements in a TBP-independent manner; these IIB-recognition elements (BREs) are localized immediately upstream (BREu), 5'-[GC][GC][GA]CGCC-3', and downstream (BREd), 5'-[GA]T[TGA][TG][GT][TG][TG]-3', of the TATA box element (PubMed:10619841, PubMed:16230532, PubMed:7675079, PubMed:9420329). Modulates transcription start site selection (PubMed:10318856). Also exhibits autoacetyltransferase activity that contributes to the activated transcription (PubMed:12931194)\",\n",
       "  'keywords': \"['3D-structure', 'Acetylation', 'Acyltransferase', 'Chromosome', 'DNA-binding', 'Host-virus interaction', 'Metal-binding', 'Nucleus', 'Phosphoprotein', 'Proteomics identification', 'Reference proteome', 'Repeat', 'Transcription', 'Transcription regulation', 'Transferase', 'Zinc', 'Zinc-finger']\",\n",
       "  'searchable_text': \"Proteins: Transcription initiation factor IIB | Function: General transcription factor that plays a role in transcription initiation by RNA polymerase II (Pol II). Involved in the pre-initiation complex (PIC) formation and Pol II recruitment at promoter DNA (PubMed:12931194, PubMed:1517211, PubMed:1876184, PubMed:1946368, PubMed:27193682, PubMed:3029109, PubMed:3818643, PubMed:7601352, PubMed:8413225, PubMed:8515820, PubMed:8516311, PubMed:8516312, PubMed:9420329). Together with the TATA box-bound TBP forms the core initiation complex and provides a bridge between TBP and the Pol II-TFIIF complex (PubMed:8413225, PubMed:8504927, PubMed:8515820, PubMed:8516311, PubMed:8516312). Released from the PIC early following the onset of transcription during the initiation and elongation transition and reassociates with TBP during the next transcription cycle (PubMed:7601352). Associates with chromatin to core promoter-specific regions (PubMed:12931194, PubMed:24441171). Binds to two distinct DNA core promoter consensus sequence elements in a TBP-independent manner; these IIB-recognition elements (BREs) are localized immediately upstream (BREu), 5'-[GC][GC][GA]CGCC-3', and downstream (BREd), 5'-[GA]T[TGA][TG][GT][TG][TG]-3', of the TATA box element (PubMed:10619841, PubMed:16230532, PubMed:7675079, PubMed:9420329). Modulates transcription start site selection (PubMed:10318856). Also exhibits autoacetyltransferase activity that contributes to the activated transcription (PubMed:12931194)\",\n",
       "  'score': 0.8538062572479248},\n",
       " {'_id': ObjectId('685ec5a12516af4a76508efb'),\n",
       "  'uniprot_id': 'O60675',\n",
       "  'protein_name': 'Transcription factor MafK',\n",
       "  'organism': 'Homo sapiens',\n",
       "  'function': 'Since they lack a putative transactivation domain, the small Mafs behave as transcriptional repressors when they dimerize among themselves (PubMed:9150357). However, they act as transcriptional activators by dimerizing with other (usually larger) basic-zipper proteins, such as NFE2, NFE2L1/NRF1, NFE2L2/NRF2 and NFE2L3/NRF3, and recruiting them to specific DNA-binding sites (PubMed:8932385, PubMed:9150357). Small Maf proteins heterodimerize with Fos and may act as competitive repressors of the NF-E2 transcription factor (PubMed:9150357)',\n",
       "  'keywords': \"['DNA-binding', 'Isopeptide bond', 'Nucleus', 'Phosphoprotein', 'Proteomics identification', 'Reference proteome', 'Repressor', 'Transcription', 'Transcription regulation', 'Ubl conjugation']\",\n",
       "  'searchable_text': 'Transcription factor MafK | Since they lack a putative transactivation domain, the small Mafs behave as transcriptional repressors when they dimerize among themselves (PubMed:9150357). However, they act as transcriptional activators by dimerizing with other (usually larger) basic-zipper proteins, such as NFE2, NFE2L1/NRF1, NFE2L2/NRF2 and NFE2L3/NRF3, and recruiting them to specific DNA-binding sites (PubMed:8932385, PubMed:9150357). Small Maf proteins heterodimerize with Fos and may act as competitive repressors of the NF-E2 transcription factor (PubMed:9150357)',\n",
       "  'score': 0.8499280214309692},\n",
       " {'_id': ObjectId('685ec5a12516af4a76508f5e'),\n",
       "  'uniprot_id': 'B2D6P4',\n",
       "  'protein_name': 'TATA box-binding protein-like 1',\n",
       "  'organism': 'Caenorhabditis elegans',\n",
       "  'function': 'May be a general transcription factor (PubMed:11030349, PubMed:11030350). Plays an essential role for RNA polymerase II/ama-1 transcription in early embryos whereby it activates a subset of RNA polymerase II promoters and facilitates the reestablishment of transcription after mitosis (PubMed:14726532)',\n",
       "  'keywords': \"['Alternative splicing', 'DNA-binding', 'Nucleus', 'Reference proteome', 'Transcription', 'Transcription regulation']\",\n",
       "  'searchable_text': 'TATA box-binding protein-like 1 | May be a general transcription factor (PubMed:11030349, PubMed:11030350). Plays an essential role for RNA polymerase II/ama-1 transcription in early embryos whereby it activates a subset of RNA polymerase II promoters and facilitates the reestablishment of transcription after mitosis (PubMed:14726532)',\n",
       "  'score': 0.8485393524169922}]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "search_vector(\"transcription factor that binds DNA\", 10, collection)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a13ebc0c-4bf1-450c-8e73-22c5d3ce3286",
   "metadata": {},
   "source": [
    "Great!! you have done your first vector search using MongoDB Atlas.\n",
    "<hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d77db562-445a-4dd4-8e58-91fdd8bc4387",
   "metadata": {},
   "source": [
    "## LAB ASSIGNMENT 4 TASK\n",
    "\n",
    "Given the requirements such as follows:\n",
    "1. User may query and find proteins on specific organism such as focus only homo sapiens\n",
    "2. GO molecular functions should be included\n",
    "3. User may query specific length of sequence.\n",
    "\n",
    "You are given the following tasks to complete:\n",
    "1. Enhance the representation of the `searchable_text`, include more information as per the requirement.\n",
    "2. Then rework on the embeddings in the MongoDB Atlas. You may delete the existing db/collection and recreate again with the new one.\n",
    "3. Re-create the search index based on the latest collection\n",
    "4. Create a simple Streamlit UI that having a simple search query box, to demonstrate the vector search of your MongoDB Atlas. Design a simple display of the information retrieved from MongoDB.\n",
    "5. As demo purpose, you should make recording to demonstrate how your streamlit app would work. (voice explanation is optional)\n",
    "6. Submit your work:\n",
    "     - revised / reworked final version of python notebook or python script. (the link to MongoDB Atlas should be working) that shows how you prepare the data and store in MongoDB Atlas and definition of the index.\n",
    "     - the streamlit app code that used for demonstrating the vector search\n",
    "       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "1deea3fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_searchable(row):\n",
    "    searchable_text = (\n",
    "        f\"Protein: {row['protein_name']} | \"\n",
    "        f\"Function: {row['function']} | \"\n",
    "        f\"Organism: {row['organism']} | \"\n",
    "        f\"GO Molecular Function: {row['go_molecular_function']} | \"\n",
    "        f\"Sequence Length: {row['sequence_length']}\"\n",
    "    )\n",
    "    return searchable_text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "05866652",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Old collection dropped.\n"
     ]
    }
   ],
   "source": [
    "collection.drop()\n",
    "print(\"Old collection dropped.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "93c8099c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing document: 349/349\n"
     ]
    }
   ],
   "source": [
    "protein_docs = []\n",
    "\n",
    "for idx, row in proteins_data.iterrows():\n",
    "    clear_output(wait=True)\n",
    "    print(f'Processing document: {idx+1}/{proteins_data.shape[0]}')\n",
    "\n",
    "    searchable_text = format_searchable(row)\n",
    "\n",
    "    doc = {\n",
    "        'uniprot_id': row['uniprot_id'],\n",
    "        'entry_name': row['entry_name'],\n",
    "        'protein_name': row['protein_name'],\n",
    "        'gene_names': row['gene_names'],\n",
    "        'organism': row['organism'],\n",
    "        'sequence_length': int(row['sequence_length']),\n",
    "        'sequence': row['sequence'],\n",
    "        'function': row['function'],\n",
    "        'subcellular_location': row['subcellular_location'],\n",
    "        'domains': row['domains'],\n",
    "        'keywords': row['keywords'],\n",
    "        'go_molecular_function': row['go_molecular_function'],\n",
    "        'go_biological_process': row['go_biological_process'],\n",
    "        'go_cellular_component': row['go_cellular_component'],\n",
    "        'searchable_text': searchable_text,\n",
    "        'embedding': ollama_embed(searchable_text)  # or st_embed(searchable_text)\n",
    "    }\n",
    "\n",
    "    protein_docs.append(doc)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "bd5be7d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_mongodb import MongoDBAtlasVectorSearch\n",
    "from langchain_ollama import OllamaEmbeddings\n",
    "\n",
    "embed_model = OllamaEmbeddings(model='nomic-embed-text')\n",
    "\n",
    "vectorstore = MongoDBAtlasVectorSearch(\n",
    "    collection=collection,\n",
    "    embedding=embed_model,\n",
    "    index_name=\"prot_index\",\n",
    "    relevance_score_fn='cosine'\n",
    ")\n",
    "\n",
    "vectorstore.create_vector_search_index(dimensions=768)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d878dcd",
   "metadata": {},
   "source": [
    "Video Link - BEN LIM CHOONG CHUEN B23CS0032\n",
    "\n",
    "https://youtu.be/ksHqlBc30xU"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
